// @algorandfoundation/puya-ts/compile-Bvr0qw3C.js
import { cV as mkDirIfNotExists, cW as PuyaService, cX as deserializeAndLog, cY as LogLevel, cO as registerPTypes, cZ as logger, c_ as createTsProgram, cS as LoggingContext, c$ as buildAwst, d0 as validateAwst, cP as typeRegistry } from './options-_3gbKwvv.js';
import { r as resolvePuyaPath, a as appVersion } from './resolve-puya-path-ORzGtNzH.js';
import ts from 'typescript';

function buildCompilationSetMapping({ awst, inputPaths, compilationSet, }) {
    const setIds = new Set(compilationSet.compilationOutputSet.map((s) => s.id));
    return awst.reduce((acc, cur) => {
        if (setIds.has(cur.id.toString())) {
            const matchedPath = inputPaths.find((p) => cur.sourceLocation.file && p.sourceFile.equals(cur.sourceLocation.file));
            if (matchedPath) {
                mkDirIfNotExists(matchedPath.outDir.toString());
                acc[cur.id.toString()] = matchedPath.outDir.toString();
            }
        }
        return acc;
    }, {});
}

async function puyaCompile({ moduleAwst, programDirectory, sourceFiles, options, compilationSet, puyaService, }) {
    const puyaPath = await resolvePuyaPath(options);
    const localPuyaService = puyaService ?? new PuyaService({ puyaPath });
    const puyaOptions = options.buildPuyaOptions(buildCompilationSetMapping({
        awst: moduleAwst,
        inputPaths: options.filePaths,
        compilationSet,
    }));
    const response = await localPuyaService.compile({
        awst: moduleAwst,
        options: puyaOptions,
        base_path: programDirectory.toString(),
        log_level: getPuyaLogLevel(options.logLevel),
        source_annotations: getSourceFileContents(sourceFiles),
    });
    // Shutdown service if we created it locally
    if (puyaService === undefined)
        await localPuyaService.shutdown();
    for (const log of response.logs) {
        deserializeAndLog(log);
    }
}
function getPuyaLogLevel(logLevel) {
    switch (logLevel) {
        case LogLevel.Debug:
            return 'debug';
        case LogLevel.Info:
            return 'info';
        case LogLevel.Warning:
            return 'warning';
        case LogLevel.Error:
            return 'error';
        case LogLevel.Critical:
            return 'critical';
    }
}
function getSourceFileContents(sourceFiles) {
    return Object.fromEntries(Object.entries(sourceFiles).map(([key, value]) => {
        const source = ts.isSourceFile(value) ? value.getFullText().replace(/\r\n/g, '\n').split(/\n/g) : value;
        return [key, source];
    }));
}

async function compile(options, puyaService) {
    const loggerCtx = LoggingContext.current;
    registerPTypes(typeRegistry);
    logger.info(undefined, appVersion({ withAVMVersion: false }));
    const programResult = createTsProgram(options);
    if (loggerCtx.hasErrors()) {
        logger.info(undefined, 'Compilation halted due to parse errors');
        return {
            programDirectory: programResult.programDirectory,
            ast: programResult.sourceFiles,
        };
    }
    const { moduleAwst, compilationSet } = buildAwst(programResult, options);
    validateAwst(moduleAwst);
    if (loggerCtx.hasErrors()) {
        logger.info(undefined, 'Compilation halted due to errors');
        return {
            programDirectory: programResult.programDirectory,
            awst: moduleAwst,
            ast: programResult.sourceFiles,
            compilationSet,
        };
    }
    if (!options.dryRun) {
        await puyaCompile({
            options,
            moduleAwst,
            programDirectory: programResult.programDirectory,
            compilationSet,
            sourceFiles: programResult.sourceFiles,
            puyaService,
        });
    }
    return {
        programDirectory: programResult.programDirectory,
        awst: moduleAwst,
        ast: programResult.sourceFiles,
        compilationSet,
    };
}

export { compile as c };
