// @algorandfoundation/puya-ts/index.mjs
import { A as ABICompatibleInstanceType, a as ALL_OP_ENUMS, b as ARC4ArrayType, c as ARC4BoolClass, d as ARC4EncodedType, e as ARC4InstanceType, f as ARC4StrClass, g as ARC4StructClass, h as ARC4StructType, i as ARC4TupleType, j as AccountFunction, k as AddressClass, l as AnyPType, m as ApplicationFunctionType, n as ApplicationTxnFunction, o as ApprovalProgram, p as Arc4TupleGeneric, q as ArrayGeneric, r as ArrayLiteralPType, s as ArrayPType, t as AssetConfigTxnFunction, u as AssetFreezeTxnFunction, v as AssetFunction, w as AssetTransferTxnFunction, B as BaseContractClassType, x as BigIntLiteralPType, y as BigIntPType, z as BigUintFunction, D as BooleanFunction, E as BoxGeneric, F as BoxMapGeneric, G as BoxMapPType, H as BoxPType, I as ByteClass, J as BytesFunction, K as BytesGeneric, L as BytesPType, M as ClassMethodDecoratorContext, N as ClearStateProgram, O as ClusteredContractClassType, P as ClusteredPrototype, Q as ContractClassPType, R as ContractProxyGeneric, S as ContractProxyType, T as DynamicArrayGeneric, U as DynamicArrayType, V as DynamicBytesConstructor, W as DynamicBytesType, X as FixedArrayGeneric, Y as FixedArrayPType, Z as FunctionPType, _ as GeneratorGeneric, $ as GeneratorType, a0 as GenericPType, a1 as GlobalStateGeneric, a2 as GlobalStateType, a3 as GroupTransactionPType, a4 as ImmutableObjectPType, a5 as InnerTransactionPType, a6 as InstanceType, a7 as InternalType, a8 as IntersectionPType, a9 as IntrinsicEnumType, aa as IntrinsicFunctionGroupType, ab as IntrinsicFunctionGroupTypeType, ac as IntrinsicFunctionType, ad as IntrinsicFunctionTypeType, ae as IterableIteratorGeneric, af as IterableIteratorType, ag as ItxnParamsPType, ah as KeyRegistrationTxnFunction, ai as LibClassType, aj as LibFunctionType, ak as LibObjType, al as LocalStateGeneric, am as LocalStateType, an as LogicSigPType, ao as MutableObjectPType, ap as MutableTuplePType, aq as NamespacePType, ar as NumberPType, as as NumericLiteralPType, at as ObjectLiteralPType, au as ObjectWithOptionalFieldsType, av as PType, aw as PaymentTxnFunction, ax as PolytypeClassMethodHelper, ay as PromiseGeneric, az as PromiseType, aA as ReadonlyArrayGeneric, aB as ReadonlyArrayPType, aC as ReadonlyGeneric, aD as ReadonlyTuplePType, aE as ReferenceArrayGeneric, aF as ReferenceArrayType, aG as StaticArrayGeneric, aH as StaticArrayType, aI as StaticBytesGeneric, aJ as StaticBytesType, aK as StringFunction, aL as SuperPrototypeSelector, aM as SuperPrototypeSelectorGeneric, aN as TemplateVarFunction, aO as TransactionFunction, aP as TransactionFunctionType, aQ as TransientType, aR as TypeParameterType, aS as TypedApplicationCallResponseGeneric, aT as TypedApplicationCallResponseType, aU as UFixedNxMGeneric, aV as UFixedNxMType, aW as Uint64EnumMemberLiteralType, aX as Uint64EnumMemberType, aY as Uint64EnumType, aZ as Uint64Function, a_ as UintN128Class, a$ as UintN16Class, b0 as UintN256Class, b1 as UintN32Class, b2 as UintN64Class, b3 as UintN8Class, b4 as UintNGeneric, b5 as UintNType, b6 as UnionPType, b7 as UnsupportedType, b8 as abiCallFunction, b9 as accountPType, ba as anyGtxnType, bb as anyItxnParamsType, bc as anyItxnType, bd as anyPType, be as applicationCallGtxnType, bf as applicationCallItxnFn, bg as applicationCallItxnParamsType, bh as applicationItxnType, bi as applicationPType, bj as arc28EmitFunction, bk as arc4AbiMethodDecorator, bl as arc4AddressAlias, bm as arc4BareMethodDecorator, bn as arc4BaseContractType, bo as arc4BooleanType, bp as arc4ByteAlias, bq as arc4StringType, br as arc4StructBaseType, bs as arc4Uint64, bt as assertFunction, bu as assertMatchFunction, bv as assetConfigGtxnType, bw as assetConfigItxnFn, bx as assetConfigItxnParamsType, by as assetConfigItxnType, bz as assetFreezeGtxnType, bA as assetFreezeItxnFn, bB as assetFreezeItxnParamsType, bC as assetFreezeItxnType, bD as assetPType, bE as assetTransferGtxnType, bF as assetTransferItxnFn, bG as assetTransferItxnParamsType, bH as assetTransferItxnType, bI as base64PType, bJ as baseContractType, bK as bigIntPType, bL as biguintPType, bM as boolPType, bN as bytesPType, bO as bzeroFunction, bP as cloneFunctionPType, bQ as compileArc4Function, bR as compileFunctionType, bS as compiledContractType, bT as compiledLogicSigType, bU as contractOptionsDecorator, bV as convertBytesFunction, bW as decodeArc4Function, bX as ecPType, bY as ecdsaPType, bZ as encodeArc4Function, b_ as ensureBudgetFunction, b$ as errFunction, c0 as esSymbol, c1 as inputOnlyObjects, c2 as isArrayType, c3 as isObjectType, c4 as isTupleLike, c5 as itoaMethod, c6 as itxnComposePType, c7 as keyRegistrationGtxnType, c8 as keyRegistrationItxnFn, c9 as keyRegistrationItxnParamsType, ca as keyRegistrationItxnType, cb as logFunction, cc as logicSigBaseType, cd as logicSigOptionsDecorator, ce as matchFunction, cf as methodSelectorFunction, cg as mimcConfigurationsPType, ch as neverPType, ci as nullPType, cj as numberPType, ck as onCompleteActionType, cl as opUpFeeSourceType, cm as paymentGtxnType, cn as paymentItxnFn, co as paymentItxnParamsType, cp as paymentItxnType, cq as ptypeToArc4EncodedType, cr as readonlyDecorator, cs as sizeOfFunction, ct as stringPType, cu as submitGroupItxnFunction, cv as transactionTypeType, cw as uint64PType, cx as undefinedPType, cy as unknownPType, cz as urangeFunction, cA as validateEncodingFunctionPType, cB as voidPType, cC as vrfVerifyPType, cD as uint8ArrayToBigInt, cE as uint8ArrayToBase64, cF as uint8ArrayToBase32, cG as uint8ArrayToHex, cH as uint8ArrayToUtf8, cI as base64ToUint8Array, cJ as base32ToUint8Array, cK as hexToUint8Array, cL as bigIntToUint8Array, cM as utf8ToUint8Array } from './options-_3gbKwvv.js';
export { cU as AbsolutePath, cT as CompileOptions, cS as LoggingContext, cN as SourceLocation, cQ as TypeResolver, cR as processInputPaths, cO as registerPTypes, cP as typeRegistry } from './options-_3gbKwvv.js';
export { c as compile } from './compile-Bvr0qw3C.js';
import 'arcsecond';
import 'node:fs';
import 'node:util';
import 'pathe';
import 'tslib';
import 'node:buffer';
import 'typescript';
import 'node:crypto';
import 'node:async_hooks';
import 'cross-spawn';
import 'util';
import 'path';
import 'os';
import 'crypto';
import 'net';
import 'change-case';
import 'polytype';
import 'glob';
import 'minimatch';
import './resolve-puya-path-ORzGtNzH.js';
import 'fs';
import 'node:url';
import 'signal-exit';
import 'tar';
import 'node:zlib';

/* This file aggregates all ptypes into a single export for external consumption of the compiler api */

var forExport = /*#__PURE__*/Object.freeze({
    __proto__: null,
    ABICompatibleInstanceType: ABICompatibleInstanceType,
    ALL_OP_ENUMS: ALL_OP_ENUMS,
    ARC4ArrayType: ARC4ArrayType,
    ARC4BoolClass: ARC4BoolClass,
    ARC4EncodedType: ARC4EncodedType,
    ARC4InstanceType: ARC4InstanceType,
    ARC4StrClass: ARC4StrClass,
    ARC4StructClass: ARC4StructClass,
    ARC4StructType: ARC4StructType,
    ARC4TupleType: ARC4TupleType,
    AccountFunction: AccountFunction,
    AddressClass: AddressClass,
    AnyPType: AnyPType,
    ApplicationFunctionType: ApplicationFunctionType,
    ApplicationTxnFunction: ApplicationTxnFunction,
    ApprovalProgram: ApprovalProgram,
    Arc4TupleGeneric: Arc4TupleGeneric,
    ArrayGeneric: ArrayGeneric,
    ArrayLiteralPType: ArrayLiteralPType,
    ArrayPType: ArrayPType,
    AssetConfigTxnFunction: AssetConfigTxnFunction,
    AssetFreezeTxnFunction: AssetFreezeTxnFunction,
    AssetFunction: AssetFunction,
    AssetTransferTxnFunction: AssetTransferTxnFunction,
    BaseContractClassType: BaseContractClassType,
    BigIntLiteralPType: BigIntLiteralPType,
    BigIntPType: BigIntPType,
    BigUintFunction: BigUintFunction,
    BooleanFunction: BooleanFunction,
    BoxGeneric: BoxGeneric,
    BoxMapGeneric: BoxMapGeneric,
    BoxMapPType: BoxMapPType,
    BoxPType: BoxPType,
    ByteClass: ByteClass,
    BytesFunction: BytesFunction,
    BytesGeneric: BytesGeneric,
    BytesPType: BytesPType,
    ClassMethodDecoratorContext: ClassMethodDecoratorContext,
    ClearStateProgram: ClearStateProgram,
    ClusteredContractClassType: ClusteredContractClassType,
    ClusteredPrototype: ClusteredPrototype,
    ContractClassPType: ContractClassPType,
    ContractProxyGeneric: ContractProxyGeneric,
    ContractProxyType: ContractProxyType,
    DynamicArrayGeneric: DynamicArrayGeneric,
    DynamicArrayType: DynamicArrayType,
    DynamicBytesConstructor: DynamicBytesConstructor,
    DynamicBytesType: DynamicBytesType,
    FixedArrayGeneric: FixedArrayGeneric,
    FixedArrayPType: FixedArrayPType,
    FunctionPType: FunctionPType,
    GeneratorGeneric: GeneratorGeneric,
    GeneratorType: GeneratorType,
    GenericPType: GenericPType,
    GlobalStateGeneric: GlobalStateGeneric,
    GlobalStateType: GlobalStateType,
    GroupTransactionPType: GroupTransactionPType,
    ImmutableObjectPType: ImmutableObjectPType,
    InnerTransactionPType: InnerTransactionPType,
    InstanceType: InstanceType,
    InternalType: InternalType,
    IntersectionPType: IntersectionPType,
    IntrinsicEnumType: IntrinsicEnumType,
    IntrinsicFunctionGroupType: IntrinsicFunctionGroupType,
    IntrinsicFunctionGroupTypeType: IntrinsicFunctionGroupTypeType,
    IntrinsicFunctionType: IntrinsicFunctionType,
    IntrinsicFunctionTypeType: IntrinsicFunctionTypeType,
    IterableIteratorGeneric: IterableIteratorGeneric,
    IterableIteratorType: IterableIteratorType,
    ItxnParamsPType: ItxnParamsPType,
    KeyRegistrationTxnFunction: KeyRegistrationTxnFunction,
    LibClassType: LibClassType,
    LibFunctionType: LibFunctionType,
    LibObjType: LibObjType,
    LocalStateGeneric: LocalStateGeneric,
    LocalStateType: LocalStateType,
    LogicSigPType: LogicSigPType,
    MutableObjectPType: MutableObjectPType,
    MutableTuplePType: MutableTuplePType,
    NamespacePType: NamespacePType,
    NumberPType: NumberPType,
    NumericLiteralPType: NumericLiteralPType,
    ObjectLiteralPType: ObjectLiteralPType,
    ObjectWithOptionalFieldsType: ObjectWithOptionalFieldsType,
    PType: PType,
    PaymentTxnFunction: PaymentTxnFunction,
    PolytypeClassMethodHelper: PolytypeClassMethodHelper,
    PromiseGeneric: PromiseGeneric,
    PromiseType: PromiseType,
    ReadonlyArrayGeneric: ReadonlyArrayGeneric,
    ReadonlyArrayPType: ReadonlyArrayPType,
    ReadonlyGeneric: ReadonlyGeneric,
    ReadonlyTuplePType: ReadonlyTuplePType,
    ReferenceArrayGeneric: ReferenceArrayGeneric,
    ReferenceArrayType: ReferenceArrayType,
    StaticArrayGeneric: StaticArrayGeneric,
    StaticArrayType: StaticArrayType,
    StaticBytesGeneric: StaticBytesGeneric,
    StaticBytesType: StaticBytesType,
    StringFunction: StringFunction,
    SuperPrototypeSelector: SuperPrototypeSelector,
    SuperPrototypeSelectorGeneric: SuperPrototypeSelectorGeneric,
    TemplateVarFunction: TemplateVarFunction,
    TransactionFunction: TransactionFunction,
    TransactionFunctionType: TransactionFunctionType,
    TransientType: TransientType,
    TypeParameterType: TypeParameterType,
    TypedApplicationCallResponseGeneric: TypedApplicationCallResponseGeneric,
    TypedApplicationCallResponseType: TypedApplicationCallResponseType,
    UFixedNxMGeneric: UFixedNxMGeneric,
    UFixedNxMType: UFixedNxMType,
    Uint64EnumMemberLiteralType: Uint64EnumMemberLiteralType,
    Uint64EnumMemberType: Uint64EnumMemberType,
    Uint64EnumType: Uint64EnumType,
    Uint64Function: Uint64Function,
    UintN128Class: UintN128Class,
    UintN16Class: UintN16Class,
    UintN256Class: UintN256Class,
    UintN32Class: UintN32Class,
    UintN64Class: UintN64Class,
    UintN8Class: UintN8Class,
    UintNGeneric: UintNGeneric,
    UintNType: UintNType,
    UnionPType: UnionPType,
    UnsupportedType: UnsupportedType,
    abiCallFunction: abiCallFunction,
    accountPType: accountPType,
    anyGtxnType: anyGtxnType,
    anyItxnParamsType: anyItxnParamsType,
    anyItxnType: anyItxnType,
    anyPType: anyPType,
    applicationCallGtxnType: applicationCallGtxnType,
    applicationCallItxnFn: applicationCallItxnFn,
    applicationCallItxnParamsType: applicationCallItxnParamsType,
    applicationItxnType: applicationItxnType,
    applicationPType: applicationPType,
    arc28EmitFunction: arc28EmitFunction,
    arc4AbiMethodDecorator: arc4AbiMethodDecorator,
    arc4AddressAlias: arc4AddressAlias,
    arc4BareMethodDecorator: arc4BareMethodDecorator,
    arc4BaseContractType: arc4BaseContractType,
    arc4BooleanType: arc4BooleanType,
    arc4ByteAlias: arc4ByteAlias,
    arc4StringType: arc4StringType,
    arc4StructBaseType: arc4StructBaseType,
    arc4Uint64: arc4Uint64,
    assertFunction: assertFunction,
    assertMatchFunction: assertMatchFunction,
    assetConfigGtxnType: assetConfigGtxnType,
    assetConfigItxnFn: assetConfigItxnFn,
    assetConfigItxnParamsType: assetConfigItxnParamsType,
    assetConfigItxnType: assetConfigItxnType,
    assetFreezeGtxnType: assetFreezeGtxnType,
    assetFreezeItxnFn: assetFreezeItxnFn,
    assetFreezeItxnParamsType: assetFreezeItxnParamsType,
    assetFreezeItxnType: assetFreezeItxnType,
    assetPType: assetPType,
    assetTransferGtxnType: assetTransferGtxnType,
    assetTransferItxnFn: assetTransferItxnFn,
    assetTransferItxnParamsType: assetTransferItxnParamsType,
    assetTransferItxnType: assetTransferItxnType,
    base64PType: base64PType,
    baseContractType: baseContractType,
    bigIntPType: bigIntPType,
    biguintPType: biguintPType,
    boolPType: boolPType,
    bytesPType: bytesPType,
    bzeroFunction: bzeroFunction,
    cloneFunctionPType: cloneFunctionPType,
    compileArc4Function: compileArc4Function,
    compileFunctionType: compileFunctionType,
    compiledContractType: compiledContractType,
    compiledLogicSigType: compiledLogicSigType,
    contractOptionsDecorator: contractOptionsDecorator,
    convertBytesFunction: convertBytesFunction,
    decodeArc4Function: decodeArc4Function,
    ecPType: ecPType,
    ecdsaPType: ecdsaPType,
    encodeArc4Function: encodeArc4Function,
    ensureBudgetFunction: ensureBudgetFunction,
    errFunction: errFunction,
    esSymbol: esSymbol,
    inputOnlyObjects: inputOnlyObjects,
    isArrayType: isArrayType,
    isObjectType: isObjectType,
    isTupleLike: isTupleLike,
    itoaMethod: itoaMethod,
    itxnComposePType: itxnComposePType,
    keyRegistrationGtxnType: keyRegistrationGtxnType,
    keyRegistrationItxnFn: keyRegistrationItxnFn,
    keyRegistrationItxnParamsType: keyRegistrationItxnParamsType,
    keyRegistrationItxnType: keyRegistrationItxnType,
    logFunction: logFunction,
    logicSigBaseType: logicSigBaseType,
    logicSigOptionsDecorator: logicSigOptionsDecorator,
    matchFunction: matchFunction,
    methodSelectorFunction: methodSelectorFunction,
    mimcConfigurationsPType: mimcConfigurationsPType,
    neverPType: neverPType,
    nullPType: nullPType,
    numberPType: numberPType,
    onCompleteActionType: onCompleteActionType,
    opUpFeeSourceType: opUpFeeSourceType,
    paymentGtxnType: paymentGtxnType,
    paymentItxnFn: paymentItxnFn,
    paymentItxnParamsType: paymentItxnParamsType,
    paymentItxnType: paymentItxnType,
    ptypeToArc4EncodedType: ptypeToArc4EncodedType,
    readonlyDecorator: readonlyDecorator,
    sizeOfFunction: sizeOfFunction,
    stringPType: stringPType,
    submitGroupItxnFunction: submitGroupItxnFunction,
    transactionTypeType: transactionTypeType,
    uint64PType: uint64PType,
    undefinedPType: undefinedPType,
    unknownPType: unknownPType,
    urangeFunction: urangeFunction,
    validateEncodingFunctionPType: validateEncodingFunctionPType,
    voidPType: voidPType,
    vrfVerifyPType: vrfVerifyPType
});

const encodingUtil = {
    utf8ToUint8Array,
    bigIntToUint8Array,
    hexToUint8Array,
    base32ToUint8Array,
    base64ToUint8Array,
    uint8ArrayToUtf8,
    uint8ArrayToHex,
    uint8ArrayToBase32,
    uint8ArrayToBase64,
    uint8ArrayToBigInt,
};

export { encodingUtil, forExport as ptypes };
