// @algorandfoundation/puya-ts/resolve-puya-path-ORzGtNzH.js
import * as fs from 'fs';
import fs__default, { rmdirSync, existsSync, statSync, mkdirSync } from 'fs';
import { fileURLToPath } from 'node:url';
import pathe__default from 'pathe';
import { C as Constants, cZ as logger, d6 as sleep, d7 as InternalError } from './options-_3gbKwvv.js';
import { __addDisposableResource, __disposeResources } from 'tslib';
import { onExit } from 'signal-exit';
import { spawn } from 'cross-spawn';
import * as require$$3 from 'crypto';
import * as path from 'path';
import * as tar from 'tar';
import { globIterateSync } from 'glob';
import { writeFileSync } from 'node:fs';
import { gzipSync } from 'node:zlib';
import require$$2 from 'os';

function packageVersion() {
    let dirName = pathe__default.dirname(fileURLToPath(import.meta.url));
    while (true) {
        const packageJsonPath = pathe__default.join(dirName, 'package.json');
        if (fs__default.existsSync(packageJsonPath)) {
            return JSON.parse(fs__default.readFileSync(packageJsonPath, 'utf8')).version;
        }
        if (dirName === pathe__default.dirname(dirName))
            break;
        dirName = pathe__default.dirname(dirName);
    }
    return 'unknown';
}
function appVersion(options) {
    const name = options?.name ?? 'puya-ts';
    const withAVMVersion = options?.withAVMVersion ?? true;
    const version = packageVersion();
    return [`${name} ${version}`]
        .concat(withAVMVersion
        ? ['', 'Targets:', `puya ${Constants.targetedPuyaVersion}`, `AVM ${Constants.supportedAvmVersions.join(', ')}`]
        : [`puya ${Constants.targetedPuyaVersion}`])
        .join('\n');
}

const lockPaths = new Set();
function acquireLock(path, options) {
    const { staleMs = 60 * 1000 } = options ?? {};
    lockPaths.add(path);
    if (existsSync(path)) {
        const now = Date.now();
        const fileStat = statSync(path);
        if (now - fileStat.mtime.getTime() > staleMs) {
            // Remove the lock file if it has expired
            rmdirSync(path);
        }
        else {
            throw new Error('Lock file already exists');
        }
    }
    mkdirSync(path, { recursive: true });
}
async function createLockFile(path, options) {
    const { maxRetries = 3, delayMs = 1000 } = options ?? {};
    for (let i = 0; i < maxRetries; i++) {
        try {
            acquireLock(path, options);
            return {
                async [Symbol.asyncDispose]() {
                    await unlockFile(path);
                },
            };
        }
        catch (err) {
            logger.debug(undefined, `Failed to acquire lock file ${path}: ${err}`);
            await sleep(delayMs);
        }
    }
    throw new Error('Failed to lock file');
}
async function unlockFile(path) {
    try {
        rmdirSync(path);
        lockPaths.delete(path);
    }
    catch (err) {
        logger.warn(undefined, `Failed to unlock file ${path}: ${err}`);
    }
}
onExit(() => {
    for (const path of lockPaths) {
        unlockFile(path);
    }
});

class InvokeCliError extends Error {
    exit;
    constructor(exit) {
        this.exit = exit;
        if (exit.code) {
            super(`Exited with code ${exit.code}`);
        }
        else if (exit.signal) {
            super(`Exited with signal ${exit.signal}`);
        }
        else {
            super('Exited with unknown cause');
        }
    }
}
function invokeCli(options) {
    return new Promise((resolve, reject) => {
        const lineAggregator = new LineAggregator(options.onReceiveLine);
        const process = spawn(options.command, options.args, {
            stdio: 'pipe',
        });
        process.stdout.on('data', (data) => lineAggregator.aggregate(data));
        process.once('close', (code) => {
            lineAggregator.flush();
            resolve({
                code: code ?? 0,
                lines: lineAggregator.lines,
            });
        });
        process.once('exit', (code, signal) => {
            if (signal !== null) {
                reject(new InvokeCliError({ signal }));
            }
            lineAggregator.flush();
            resolve({
                code: code ?? 0,
                lines: lineAggregator.lines,
            });
        });
        process.once('error', reject);
    });
}
class LineAggregator {
    onLine;
    #line = '';
    #lines = [];
    constructor(onLine) {
        this.onLine = onLine;
    }
    aggregate(chunk) {
        if (chunk === undefined || chunk === null)
            return;
        const text = chunk.toString('utf-8');
        for (const c of text) {
            switch (c) {
                case '\n':
                    this.flushLine();
                    break;
                case '\r':
                    continue;
                default:
                    this.#line += c;
                    break;
            }
        }
    }
    flush() {
        if (this.#line)
            this.flushLine();
    }
    flushLine() {
        this.#lines.push(this.#line);
        this.onLine?.(this.#line);
        this.#line = '';
    }
    get lines() {
        return this.#lines;
    }
}

async function runPuya({ command, args, cwd, onOutput, }) {
    const result = await invokeCli({
        command,
        args,
        onReceiveLine: onOutput});
    if (result.code !== 0) {
        logger.critical(undefined, `Compilation exited with status ${result.code}`);
    }
}

var VersionCompareVerdict;
(function (VersionCompareVerdict) {
    VersionCompareVerdict["ExactMatch"] = "ExactMatch";
    VersionCompareVerdict["Inconclusive"] = "Inconclusive";
    VersionCompareVerdict["MajorMismatch"] = "MajorMismatch";
    VersionCompareVerdict["MinorMismatch"] = "MinorMismatch";
    VersionCompareVerdict["OlderRevision"] = "OlderRevision";
    VersionCompareVerdict["NewerRevision"] = "NewerRevision";
})(VersionCompareVerdict || (VersionCompareVerdict = {}));
async function comparePuyaVersion(puyaPath) {
    const target = Constants.targetedPuyaVersion;
    const versionParser = new VersionParser();
    await runPuya({
        command: puyaPath,
        args: ['--version'],
        onOutput: (line) => versionParser.receiveLine(line),
    });
    if (!versionParser.version)
        return { target, verdict: VersionCompareVerdict.Inconclusive };
    const ver = versionParser.version;
    // Compare
    const [major, minor, rev] = target.split('.').map((x) => Number(x));
    if (ver.major !== major)
        return { verdict: VersionCompareVerdict.MajorMismatch, target, found: ver.formatted };
    if (ver.minor !== minor)
        return { verdict: VersionCompareVerdict.MinorMismatch, target, found: ver.formatted };
    if (ver.rev < rev)
        return { verdict: VersionCompareVerdict.OlderRevision, target, found: ver.formatted };
    if (ver.rev > rev)
        return { verdict: VersionCompareVerdict.NewerRevision, target, found: ver.formatted };
    return { verdict: VersionCompareVerdict.ExactMatch, target, found: ver.formatted };
}
async function checkPuyaVersion(puyaPath) {
    const result = await comparePuyaVersion(puyaPath);
    switch (result.verdict) {
        case VersionCompareVerdict.Inconclusive:
            logger.warn(undefined, `Unable to verify the version of Puya at ${puyaPath}.`);
            break;
        case VersionCompareVerdict.MajorMismatch:
        case VersionCompareVerdict.MinorMismatch:
            logger.warn(undefined, `Version of Puya at ${puyaPath} (${result.found}) does not match targeted version (${result.target}). There may be compatibility issues.`);
            break;
        case VersionCompareVerdict.OlderRevision:
            logger.warn(undefined, `Revision of Puya at ${puyaPath} (${result.found}) is older than the targeted revision (${result.target})`);
            break;
        case VersionCompareVerdict.NewerRevision:
            logger.debug(undefined, `Revision of Puya at ${puyaPath} (${result.found}) is newer than the targeted revision (${result.target})`);
            break;
    }
}
class VersionParser {
    #ver;
    receiveLine(line) {
        const matched = /^puya ((\d+)\.(\d+)\.(\d+)(-.*)?)$/.exec(line);
        if (!matched) {
            logger.debug(undefined, `'puya --version' command returned unexpected output: "${line}"`);
        }
        else {
            this.#ver = {
                formatted: matched[1],
                major: Number(matched[2]),
                minor: Number(matched[3]),
                rev: Number(matched[4]),
                suffix: matched[5] ?? '',
            };
        }
    }
    get version() {
        return this.#ver;
    }
}

function generateTempDir() {
    const dirPath = fs__default.mkdtempSync(pathe__default.join(require$$2.tmpdir(), 'puya-ts-'));
    return {
        get dirPath() {
            return dirPath;
        },
        *files() {
            for (const p of globIterateSync(pathe__default.join(dirPath, '**'), {
                nodir: true,
            })) {
                yield p;
            }
        },
        makeFile({ name, ext, compress }) {
            const path = pathe__default.join(this.dirPath, `${name}.${ext ?? 'tmp'}${compress ? '.gz' : ''}`);
            return {
                get filePath() {
                    return path;
                },
                writeFileSync(data, options) {
                    if (compress) {
                        data = gzipSync(data);
                    }
                    writeFileSync(path, data, options);
                },
            };
        },
        [Symbol.dispose]() {
            fs__default.rmSync(dirPath, { recursive: true, force: true });
        },
    };
}

/**
 * Gets the platform-specific binary name
 * @returns The appropriate binary name for the current platform
 */
function getBinaryName() {
    return process.platform === 'win32' ? 'puya.exe' : 'puya';
}
function getPuyaTsDir() {
    try {
        return pathe__default.dirname(require.resolve('@algorandfoundation/puya-ts'));
    }
    catch {
        return pathe__default.dirname(fileURLToPath(import.meta.url));
    }
}
function getPuyaStorageDir() {
    const puyaTsDirName = getPuyaTsDir();
    return path.join(puyaTsDirName, '.puya');
}
function getCachedPuyaBinaryPath(puyaStorageDir, version) {
    const binaryFileName = getBinaryName();
    return path.join(puyaStorageDir, version.formatted, binaryFileName);
}
/**
 * Gets the platform-specific details for downloads
 * @returns Object containing OS and architecture information
 */
function getPlatformDetails() {
    const cliMessage = `Please refer to the puya-ts CLI documentation for information on how to install the Puya backend manually.`;
    // Map Node.js platform to OS name used in filenames
    let os;
    switch (process.platform) {
        case 'win32':
            os = 'windows';
            break;
        case 'darwin':
            os = 'macos';
            break;
        case 'linux':
            os = 'linux';
            break;
        default:
            throw new InternalError(`Unsupported platform: ${process.platform}. ${cliMessage}`);
    }
    // Map Node.js architecture to architecture name used in filenames
    let arch;
    switch (process.arch) {
        case 'x64':
            arch = 'x64';
            break;
        case 'arm64':
            arch = 'arm64';
            break;
        default:
            throw new InternalError(`Unsupported architecture: ${process.arch}. ${cliMessage}`);
    }
    return { os, arch };
}
/**
 * Downloads the Puya binary for a specific release version
 * @param puyaStorageDir The directory where puya binaries should be stored once downloaded
 * @param version The release version to download (e.g., "1.0.0")
 * @returns Promise that resolves to the path of the extracted binary
 */
async function downloadPuyaBinary(puyaStorageDir, version) {
    const env_1 = { stack: [], error: void 0, hasError: false };
    try {
        // Get platform-specific details
        const { os, arch } = getPlatformDetails();
        // Build platform-specific filenames
        const platformId = `${os}_${arch}`;
        const archiveFileName = `puya-${version.formatted}-${platformId}.tar.gz`;
        const checksumFileName = `puya-${version.formatted}-${platformId}.sha256.txt`;
        const binaryFileName = getBinaryName();
        const extractedDir = path.join(puyaStorageDir, version.formatted);
        // Ensure our storage directories exist
        if (!fs.existsSync(extractedDir)) {
            fs.mkdirSync(extractedDir, { recursive: true });
        }
        const tempDir = __addDisposableResource(env_1, generateTempDir(), false);
        const archiveFilePath = path.join(tempDir.dirPath, archiveFileName);
        const checksumFilePath = path.join(tempDir.dirPath, checksumFileName);
        const extractedBinaryPath = path.join(extractedDir, binaryFileName);
        logger.debug(undefined, `Downloading Puya binary for version ${version.formatted} and platform ${platformId}`);
        const archiveUrl = `https://github.com/${Constants.puyaGithubRepo}/releases/download/v${version.formatted}/${archiveFileName}`;
        const checksumUrl = `https://github.com/${Constants.puyaGithubRepo}/releases/download/v${version.formatted}/${checksumFileName}`;
        await downloadFile(archiveUrl, archiveFilePath);
        await downloadFile(checksumUrl, checksumFilePath);
        await verifyChecksum(archiveFilePath, archiveFileName, checksumFilePath);
        await tar.extract({
            file: archiveFilePath,
            cwd: extractedDir,
        });
        // Check if extraction was successful and binary exists
        if (!fs.existsSync(extractedBinaryPath)) {
            throw new InternalError(`Binary file ${binaryFileName} not found in the extracted archive`);
        }
        await smokeTestBinary(extractedBinaryPath);
        logger.debug(undefined, `Successfully downloaded and extracted Puya binary to ${extractedBinaryPath}`);
        return extractedBinaryPath;
    }
    catch (e_1) {
        env_1.error = e_1;
        env_1.hasError = true;
    }
    finally {
        __disposeResources(env_1);
    }
}
async function smokeTestBinary(path) {
    logger.debug(undefined, `Smoke testing downloaded binary @ ${path}`);
    let output = '';
    await runPuya({
        command: path,
        args: ['--version'],
        onOutput: (line) => (output += line),
    });
    if (!/^puya ((\d+)\.(\d+)\.(\d+)(-.*)?)$/.test(output)) {
        throw new InternalError(`Unexpected output from puya --version : ${output}`);
    }
}
/**
 * Downloads a file from a URL to a specified path
 * @param url The URL to download from
 * @param destination The local path to save the file to
 * @returns Promise that resolves when the download is complete
 */
async function downloadFile(url, destination) {
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(`Failed to download file. Status Code: ${response.status}`);
    }
    const buffer = await response.arrayBuffer();
    return await new Promise((resolve, reject) => {
        const fileStream = fs.createWriteStream(destination);
        const nodeBuffer = Buffer.from(buffer);
        fileStream.write(nodeBuffer);
        fileStream.end();
        fileStream.on('finish', () => {
            resolve();
        });
        fileStream.on('error', (err) => {
            fs.unlinkSync(destination);
            reject(err);
        });
    });
}
/**
 * Verifies the checksum of a downloaded file
 * @param filePath Path to the file to verify
 * @param checksumFilePath Path to the checksum file
 * @returns Promise that resolves to true if the checksum matches, false otherwise
 */
async function verifyChecksum(filePath, fileName, checksumFilePath) {
    return new Promise((resolve, reject) => {
        // Read the checksum file
        const expectedChecksum = fs.readFileSync(checksumFilePath, 'utf8').trim().toLowerCase();
        // Calculate the SHA256 checksum of the downloaded file
        const hash = require$$3.createHash('sha256');
        const stream = fs.createReadStream(filePath);
        stream.on('data', (data) => {
            hash.update(data);
        });
        stream.on('end', () => {
            const calculatedChecksum = hash.digest('hex').toLowerCase();
            if (`${calculatedChecksum}  ${fileName}` === expectedChecksum) {
                resolve();
            }
            reject(new InternalError(`Checksum verification failed. Expected checksum: ${expectedChecksum} but got: ${calculatedChecksum}`));
        });
    });
}

function parseSemVer(version) {
    const matched = /^(\d+)\.(\d+)\.(\d+)(-.*)?$/.exec(version);
    if (!matched) {
        throw new InternalError(`Invalid version string: ${version}`);
    }
    return {
        major: Number(matched[1]),
        minor: Number(matched[2]),
        rev: Number(matched[3]),
        suffix: matched[4] ?? '',
        formatted: version,
    };
}

async function resolvePuyaPath(options) {
    const env_1 = { stack: [], error: void 0, hasError: false };
    try {
        if (options.customPuyaPath) {
            if (!options.skipVersionCheck) {
                await checkPuyaVersion(options.customPuyaPath);
            }
            return options.customPuyaPath;
        }
        const version = parseSemVer(Constants.targetedPuyaVersion);
        const puyaStorageDir = getPuyaStorageDir();
        const cachedBinaryPath = getCachedPuyaBinaryPath(puyaStorageDir, version);
        if (isBinaryCached(cachedBinaryPath)) {
            return cachedBinaryPath;
        }
        const _ = __addDisposableResource(env_1, await createLockFile(`${puyaStorageDir}.lock`, { maxRetries: 30, delayMs: 1000, staleMs: 60 * 1000 })
        // Between the first check and acquiring the lock, the binary may have been downloaded by another process.
        // In this case there is no need to download the binary again.
        , true);
        // Between the first check and acquiring the lock, the binary may have been downloaded by another process.
        // In this case there is no need to download the binary again.
        if (isBinaryCached(cachedBinaryPath)) {
            return cachedBinaryPath;
        }
        return await downloadPuyaBinary(puyaStorageDir, version);
    }
    catch (e_1) {
        env_1.error = e_1;
        env_1.hasError = true;
    }
    finally {
        const result_1 = __disposeResources(env_1);
        if (result_1)
            await result_1;
    }
}
function isBinaryCached(cachePath) {
    if (fs.existsSync(cachePath)) {
        logger.debug(undefined, `Found cached Puya binary at ${cachePath}`);
        return true;
    }
    return false;
}

export { appVersion as a, packageVersion as b, parseSemVer as p, resolvePuyaPath as r };
