// @algorandfoundation/algorand-typescript/op-BUFe04-Y.js
import { N as NoImplementation } from './errors-Bg4n2Chz.js';

/* THIS FILE IS GENERATED BY ~/scripts/generate-op-funcs.ts - DO NOT MODIFY DIRECTLY */
var Base64;
(function (Base64) {
    Base64["URLEncoding"] = "URLEncoding";
    Base64["StdEncoding"] = "StdEncoding";
})(Base64 || (Base64 = {}));
var Ec;
(function (Ec) {
    /**
     * G1 of the BN254 curve. Points encoded as 32 byte X following by 32 byte Y
     */
    Ec["BN254g1"] = "BN254g1";
    /**
     * G2 of the BN254 curve. Points encoded as 64 byte X following by 64 byte Y
     */
    Ec["BN254g2"] = "BN254g2";
    /**
     * G1 of the BLS 12-381 curve. Points encoded as 48 byte X following by 48 byte Y
     */
    Ec["BLS12_381g1"] = "BLS12_381g1";
    /**
     * G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 96 byte Y
     */
    Ec["BLS12_381g2"] = "BLS12_381g2";
})(Ec || (Ec = {}));
var Ecdsa;
(function (Ecdsa) {
    /**
     * secp256k1 curve, used in Bitcoin
     */
    Ecdsa["Secp256k1"] = "Secp256k1";
    /**
     * secp256r1 curve, NIST standard
     */
    Ecdsa["Secp256r1"] = "Secp256r1";
})(Ecdsa || (Ecdsa = {}));
var MimcConfigurations;
(function (MimcConfigurations) {
    /**
     * MiMC configuration for the BN254 curve with Miyaguchi-Preneel mode, 110 rounds, exponent 5, seed "seed"
     */
    MimcConfigurations["BN254Mp110"] = "BN254Mp110";
    /**
     * MiMC configuration for the BLS12-381 curve with Miyaguchi-Preneel mode, 111 rounds, exponent 5, seed "seed"
     */
    MimcConfigurations["BLS12_381Mp111"] = "BLS12_381Mp111";
})(MimcConfigurations || (MimcConfigurations = {}));
var VrfVerify;
(function (VrfVerify) {
    VrfVerify["VrfAlgorand"] = "VrfAlgorand";
})(VrfVerify || (VrfVerify = {}));
const AcctParams = {
    /**
     * Account balance in microalgos
     * Min AVM version: 6
     */
    acctBalance(a) {
        throw new NoImplementation();
    },
    /**
     * Minimum required balance for account, in microalgos
     * Min AVM version: 6
     */
    acctMinBalance(a) {
        throw new NoImplementation();
    },
    /**
     * Address the account is rekeyed to.
     * Min AVM version: 6
     */
    acctAuthAddr(a) {
        throw new NoImplementation();
    },
    /**
     * The total number of uint64 values allocated by this account in Global and Local States.
     * Min AVM version: 8
     */
    acctTotalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * The total number of byte array values allocated by this account in Global and Local States.
     * Min AVM version: 8
     */
    acctTotalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * The number of extra app code pages used by this account.
     * Min AVM version: 8
     */
    acctTotalExtraAppPages(a) {
        throw new NoImplementation();
    },
    /**
     * The number of existing apps created by this account.
     * Min AVM version: 8
     */
    acctTotalAppsCreated(a) {
        throw new NoImplementation();
    },
    /**
     * The number of apps this account is opted into.
     * Min AVM version: 8
     */
    acctTotalAppsOptedIn(a) {
        throw new NoImplementation();
    },
    /**
     * The number of existing ASAs created by this account.
     * Min AVM version: 8
     */
    acctTotalAssetsCreated(a) {
        throw new NoImplementation();
    },
    /**
     * The numbers of ASAs held by this account (including ASAs this account created).
     * Min AVM version: 8
     */
    acctTotalAssets(a) {
        throw new NoImplementation();
    },
    /**
     * The number of existing boxes created by this account's app.
     * Min AVM version: 8
     */
    acctTotalBoxes(a) {
        throw new NoImplementation();
    },
    /**
     * The total number of bytes used by this account's app's box keys and values.
     * Min AVM version: 8
     */
    acctTotalBoxBytes(a) {
        throw new NoImplementation();
    },
    /**
     * Has this account opted into block payouts
     * Min AVM version: 11
     */
    acctIncentiveEligible(a) {
        throw new NoImplementation();
    },
    /**
     * The round number of the last block this account proposed.
     * Min AVM version: 11
     */
    acctLastProposed(a) {
        throw new NoImplementation();
    },
    /**
     * The round number of the last block this account sent a heartbeat.
     * Min AVM version: 11
     */
    acctLastHeartbeat(a) {
        throw new NoImplementation();
    },
};
/**
 * A plus B as a 128-bit result. X is the carry-bit, Y is the low-order 64 bits.
 * @see Native TEAL opcode: [`addw`](https://dev.algorand.co/reference/algorand-teal/opcodes#addw)
 * Min AVM version: 2
 */
function addw(a, b) {
    throw new NoImplementation();
}
/**
 * Get or modify Global app state
 */
const AppGlobal = {
    /**
     * delete key A from the global state of the current application
     * @param a state key.
     * Deleting a key which is already absent has no effect on the application global state. (In particular, it does _not_ cause the program to fail.)
     * @see Native TEAL opcode: [`app_global_del`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_del)
     * Min AVM version: 2
     */
    delete(a) {
        throw new NoImplementation();
    },
    /**
     * global state of the key A in the current application
     * @param a state key.
     * @return value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_global_get`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_get)
     * Min AVM version: 2
     */
    getBytes(a) {
        throw new NoImplementation();
    },
    /**
     * global state of the key A in the current application
     * @param a state key.
     * @return value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_global_get`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_get)
     * Min AVM version: 2
     */
    getUint64(a) {
        throw new NoImplementation();
    },
    /**
     * X is the global state of application A, key B. Y is 1 if key existed, else 0
     * @param a Txn.ForeignApps offset (or, since v4, an _available_ application id), state key.
     * @return did_exist flag (top of the stack, 1 if the application and key existed and 0 otherwise), value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_global_get_ex`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_get_ex)
     * Min AVM version: 2
     */
    getExBytes(a, b) {
        throw new NoImplementation();
    },
    /**
     * X is the global state of application A, key B. Y is 1 if key existed, else 0
     * @param a Txn.ForeignApps offset (or, since v4, an _available_ application id), state key.
     * @return did_exist flag (top of the stack, 1 if the application and key existed and 0 otherwise), value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_global_get_ex`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_get_ex)
     * Min AVM version: 2
     */
    getExUint64(a, b) {
        throw new NoImplementation();
    },
    /**
     * write B to key A in the global state of the current application
     * @see Native TEAL opcode: [`app_global_put`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_global_put)
     * Min AVM version: 2
     */
    put(a, b) {
        throw new NoImplementation();
    },
};
/**
 * Get or modify Local app state
 */
const AppLocal = {
    /**
     * delete key B from account A's local state of the current application
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), state key.
     * Deleting a key which is already absent has no effect on the application local state. (In particular, it does _not_ cause the program to fail.)
     * @see Native TEAL opcode: [`app_local_del`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_del)
     * Min AVM version: 2
     */
    delete(a, b) {
        throw new NoImplementation();
    },
    /**
     * local state of the key B in the current application in account A
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), state key.
     * @return value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_local_get`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_get)
     * Min AVM version: 2
     */
    getBytes(a, b) {
        throw new NoImplementation();
    },
    /**
     * local state of the key B in the current application in account A
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), state key.
     * @return value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_local_get`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_get)
     * Min AVM version: 2
     */
    getUint64(a, b) {
        throw new NoImplementation();
    },
    /**
     * X is the local state of application B, key C in account A. Y is 1 if key existed, else 0
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), _available_ application id (or, since v4, a Txn.ForeignApps offset), state key.
     * @return did_exist flag (top of the stack, 1 if the application and key existed and 0 otherwise), value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_local_get_ex`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_get_ex)
     * Min AVM version: 2
     */
    getExBytes(a, b, c) {
        throw new NoImplementation();
    },
    /**
     * X is the local state of application B, key C in account A. Y is 1 if key existed, else 0
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), _available_ application id (or, since v4, a Txn.ForeignApps offset), state key.
     * @return did_exist flag (top of the stack, 1 if the application and key existed and 0 otherwise), value. The value is zero (of type uint64) if the key does not exist.
     * @see Native TEAL opcode: [`app_local_get_ex`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_get_ex)
     * Min AVM version: 2
     */
    getExUint64(a, b, c) {
        throw new NoImplementation();
    },
    /**
     * write C to key B in account A's local state of the current application
     * @param a Txn.Accounts offset (or, since v4, an _available_ account address), state key, value.
     * @see Native TEAL opcode: [`app_local_put`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_local_put)
     * Min AVM version: 2
     */
    put(a, b, c) {
        throw new NoImplementation();
    },
};
/**
 * 1 if account A is opted in to application B, else 0
 * @param a Txn.Accounts offset (or, since v4, an _available_ account address), _available_ application id (or, since v4, a Txn.ForeignApps offset).
 * @return 1 if opted in and 0 otherwise.
 * @see Native TEAL opcode: [`app_opted_in`](https://dev.algorand.co/reference/algorand-teal/opcodes#app_opted_in)
 * Min AVM version: 2
 */
function appOptedIn(a, b) {
    throw new NoImplementation();
}
const AppParams = {
    /**
     * Bytecode of Approval Program
     * Min AVM version: 5
     */
    appApprovalProgram(a) {
        throw new NoImplementation();
    },
    /**
     * Bytecode of Clear State Program
     * Min AVM version: 5
     */
    appClearStateProgram(a) {
        throw new NoImplementation();
    },
    /**
     * Number of uint64 values allowed in Global State
     * Min AVM version: 5
     */
    appGlobalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of byte array values allowed in Global State
     * Min AVM version: 5
     */
    appGlobalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of uint64 values allowed in Local State
     * Min AVM version: 5
     */
    appLocalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of byte array values allowed in Local State
     * Min AVM version: 5
     */
    appLocalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Extra Program Pages of code space
     * Min AVM version: 5
     */
    appExtraProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Creator address
     * Min AVM version: 5
     */
    appCreator(a) {
        throw new NoImplementation();
    },
    /**
     * Address for which this application has authority
     * Min AVM version: 5
     */
    appAddress(a) {
        throw new NoImplementation();
    },
    /**
     * Version of the app, incremented each time the approval or clear program changes
     * Min AVM version: 12
     */
    appVersion(a) {
        throw new NoImplementation();
    },
};
/**
 * Ath LogicSig argument
 * @see Native TEAL opcode: [`args`](https://dev.algorand.co/reference/algorand-teal/opcodes#args)
 * Min AVM version: 5
 */
function arg(a) {
    throw new NoImplementation();
}
const AssetHolding = {
    /**
     * Amount of the asset unit held by this account
     * Min AVM version: 2
     */
    assetBalance(a, b) {
        throw new NoImplementation();
    },
    /**
     * Is the asset frozen or not
     * Min AVM version: 2
     */
    assetFrozen(a, b) {
        throw new NoImplementation();
    },
};
const AssetParams = {
    /**
     * Total number of units of this asset
     * Min AVM version: 2
     */
    assetTotal(a) {
        throw new NoImplementation();
    },
    /**
     * See AssetParams.Decimals
     * Min AVM version: 2
     */
    assetDecimals(a) {
        throw new NoImplementation();
    },
    /**
     * Frozen by default or not
     * Min AVM version: 2
     */
    assetDefaultFrozen(a) {
        throw new NoImplementation();
    },
    /**
     * Asset unit name
     * Min AVM version: 2
     */
    assetUnitName(a) {
        throw new NoImplementation();
    },
    /**
     * Asset name
     * Min AVM version: 2
     */
    assetName(a) {
        throw new NoImplementation();
    },
    /**
     * URL with additional info about the asset
     * Min AVM version: 2
     */
    assetUrl(a) {
        throw new NoImplementation();
    },
    /**
     * Arbitrary commitment
     * Min AVM version: 2
     */
    assetMetadataHash(a) {
        throw new NoImplementation();
    },
    /**
     * Manager address
     * Min AVM version: 2
     */
    assetManager(a) {
        throw new NoImplementation();
    },
    /**
     * Reserve address
     * Min AVM version: 2
     */
    assetReserve(a) {
        throw new NoImplementation();
    },
    /**
     * Freeze address
     * Min AVM version: 2
     */
    assetFreeze(a) {
        throw new NoImplementation();
    },
    /**
     * Clawback address
     * Min AVM version: 2
     */
    assetClawback(a) {
        throw new NoImplementation();
    },
    /**
     * Creator address
     * Min AVM version: 5
     */
    assetCreator(a) {
        throw new NoImplementation();
    },
};
/**
 * balance for account A, in microalgos. The balance is observed after the effects of previous transactions in the group, and after the fee for the current transaction is deducted. Changes caused by inner transactions are observable immediately following `itxn_submit`
 * @param a Txn.Accounts offset (or, since v4, an _available_ account address), _available_ application id (or, since v4, a Txn.ForeignApps offset).
 * @return value.
 * @see Native TEAL opcode: [`balance`](https://dev.algorand.co/reference/algorand-teal/opcodes#balance)
 * Min AVM version: 2
 */
function balance(a) {
    throw new NoImplementation();
}
/**
 * decode A which was base64-encoded using _encoding_ E. Fail if A is not base64 encoded with encoding E
 * *Warning*: Usage should be restricted to very rare use cases. In almost all cases, smart contracts should directly handle non-encoded byte-strings.	This opcode should only be used in cases where base64 is the only available option, e.g. interoperability with a third-party that only signs base64 strings.
 *  Decodes A using the base64 encoding E. Specify the encoding with an immediate arg either as URL and Filename Safe (`URLEncoding`) or Standard (`StdEncoding`). See [RFC 4648 sections 4 and 5](https://rfc-editor.org/rfc/rfc4648.html#section-4). It is assumed that the encoding ends with the exact number of `=` padding characters as required by the RFC. When padding occurs, any unused pad bits in the encoding must be set to zero or the decoding will fail. The special cases of `\n` and `\r` are allowed but completely ignored. An error will result when attempting to decode a string with a character that is not in the encoding alphabet or not one of `=`, `\r`, or `\n`.
 * @see Native TEAL opcode: [`base64_decode`](https://dev.algorand.co/reference/algorand-teal/opcodes#base64_decode)
 * Min AVM version: 7
 */
function base64Decode(e, a) {
    throw new NoImplementation();
}
/**
 * The highest set bit in A. If A is a byte-array, it is interpreted as a big-endian unsigned integer. bitlen of 0 is 0, bitlen of 8 is 4
 * bitlen interprets arrays as big-endian integers, unlike setbit/getbit
 * @see Native TEAL opcode: [`bitlen`](https://dev.algorand.co/reference/algorand-teal/opcodes#bitlen)
 * Min AVM version: 4
 */
function bitLength(a) {
    throw new NoImplementation();
}
const Block = {
    blkSeed(a) {
        throw new NoImplementation();
    },
    blkTimestamp(a) {
        throw new NoImplementation();
    },
    blkProposer(a) {
        throw new NoImplementation();
    },
    blkFeesCollected(a) {
        throw new NoImplementation();
    },
    blkBonus(a) {
        throw new NoImplementation();
    },
    blkBranch(a) {
        throw new NoImplementation();
    },
    blkFeeSink(a) {
        throw new NoImplementation();
    },
    blkProtocol(a) {
        throw new NoImplementation();
    },
    blkTxnCounter(a) {
        throw new NoImplementation();
    },
    blkProposerPayout(a) {
        throw new NoImplementation();
    },
};
/**
 * Get or modify box state
 */
const Box = {
    /**
     * create a box named A, of length B. Fail if the name A is empty or B exceeds 32,768. Returns 0 if A already existed, else 1
     * Newly created boxes are filled with 0 bytes. `box_create` will fail if the referenced box already exists with a different size. Otherwise, existing boxes are unchanged by `box_create`.
     * @see Native TEAL opcode: [`box_create`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_create)
     * Min AVM version: 8
     */
    create(a, b) {
        throw new NoImplementation();
    },
    /**
     * delete box named A if it exists. Return 1 if A existed, 0 otherwise
     * @see Native TEAL opcode: [`box_del`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_del)
     * Min AVM version: 8
     */
    delete(a) {
        throw new NoImplementation();
    },
    /**
     * read C bytes from box A, starting at offset B. Fail if A does not exist, or the byte range is outside A's size.
     * @see Native TEAL opcode: [`box_extract`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_extract)
     * Min AVM version: 8
     */
    extract(a, b, c) {
        throw new NoImplementation();
    },
    /**
     * X is the contents of box A if A exists, else ''. Y is 1 if A exists, else 0.
     * For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `box_replace`
     * @see Native TEAL opcode: [`box_get`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_get)
     * Min AVM version: 8
     */
    get(a) {
        throw new NoImplementation();
    },
    /**
     * X is the length of box A if A exists, else 0. Y is 1 if A exists, else 0.
     * @see Native TEAL opcode: [`box_len`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_len)
     * Min AVM version: 8
     */
    length(a) {
        throw new NoImplementation();
    },
    /**
     * replaces the contents of box A with byte-array B. Fails if A exists and len(B) != len(box A). Creates A if it does not exist
     * For boxes that exceed 4,096 bytes, consider `box_create`, `box_extract`, and `box_replace`
     * @see Native TEAL opcode: [`box_put`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_put)
     * Min AVM version: 8
     */
    put(a, b) {
        throw new NoImplementation();
    },
    /**
     * write byte-array C into box A, starting at offset B. Fail if A does not exist, or the byte range is outside A's size.
     * @see Native TEAL opcode: [`box_replace`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_replace)
     * Min AVM version: 8
     */
    replace(a, b, c) {
        throw new NoImplementation();
    },
    /**
     * change the size of box named A to be of length B, adding zero bytes to end or removing bytes from the end, as needed. Fail if the name A is empty, A is not an existing box, or B exceeds 32,768.
     * @see Native TEAL opcode: [`box_resize`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_resize)
     * Min AVM version: 10
     */
    resize(a, b) {
        throw new NoImplementation();
    },
    /**
     * set box A to contain its previous bytes up to index B, followed by D, followed by the original bytes of A that began at index B+C.
     * Boxes are of constant length. If C < len(D), then len(D)-C bytes will be removed from the end. If C > len(D), zero bytes will be appended to the end to reach the box length.
     * @see Native TEAL opcode: [`box_splice`](https://dev.algorand.co/reference/algorand-teal/opcodes#box_splice)
     * Min AVM version: 10
     */
    splice(a, b, c, d) {
        throw new NoImplementation();
    },
};
/**
 * The largest integer I such that I^2 <= A. A and I are interpreted as big-endian unsigned integers
 * @see Native TEAL opcode: [`bsqrt`](https://dev.algorand.co/reference/algorand-teal/opcodes#bsqrt)
 * Min AVM version: 6
 */
function bsqrt(a) {
    throw new NoImplementation();
}
/**
 * converts big-endian byte array A to uint64. Fails if len(A) > 8. Padded by leading 0s if len(A) < 8.
 * `btoi` fails if the input is longer than 8 bytes.
 * @see Native TEAL opcode: [`btoi`](https://dev.algorand.co/reference/algorand-teal/opcodes#btoi)
 * Min AVM version: 1
 */
function btoi(a) {
    throw new NoImplementation();
}
/**
 * zero filled byte-array of length A
 * @see Native TEAL opcode: [`bzero`](https://dev.algorand.co/reference/algorand-teal/opcodes#bzero)
 * Min AVM version: 4
 */
function bzero(a) {
    throw new NoImplementation();
}
/**
 * join A and B
 * `concat` fails if the result would be greater than 4096 bytes.
 * @see Native TEAL opcode: [`concat`](https://dev.algorand.co/reference/algorand-teal/opcodes#concat)
 * Min AVM version: 2
 */
function concat(a, b) {
    throw new NoImplementation();
}
/**
 * W,X = (A,B / C,D); Y,Z = (A,B modulo C,D)
 * The notation J,K indicates that two uint64 values J and K are interpreted as a uint128 value, with J as the high uint64 and K the low.
 * @see Native TEAL opcode: [`divmodw`](https://dev.algorand.co/reference/algorand-teal/opcodes#divmodw)
 * Min AVM version: 4
 */
function divmodw(a, b, c, d) {
    throw new NoImplementation();
}
/**
 * A,B / C. Fail if C == 0 or if result overflows.
 * The notation A,B indicates that A and B are interpreted as a uint128 value, with A as the high uint64 and B the low.
 * @see Native TEAL opcode: [`divw`](https://dev.algorand.co/reference/algorand-teal/opcodes#divw)
 * Min AVM version: 6
 */
function divw(a, b, c) {
    throw new NoImplementation();
}
/**
 * Elliptic Curve functions
 */
const EllipticCurve = {
    /**
     * for curve points A and B, return the curve point A + B
     * A and B are curve points in affine representation: field element X concatenated with field element Y. Field element `Z` is encoded as follows.
     * For the base field elements (Fp), `Z` is encoded as a big-endian number and must be lower than the field modulus.
     * For the quadratic field extension (Fp2), `Z` is encoded as the concatenation of the individual encoding of the coefficients. For an Fp2 element of the form `Z = Z0 + Z1 i`, where `i` is a formal quadratic non-residue, the encoding of Z is the concatenation of the encoding of `Z0` and `Z1` in this order. (`Z0` and `Z1` must be less than the field modulus).
     * The point at infinity is encoded as `(X,Y) = (0,0)`.
     * Groups G1 and G2 are denoted additively.
     * Fails if A or B is not in G.
     * A and/or B are allowed to be the point at infinity.
     * Does _not_ check if A and B are in the main prime-order subgroup.
     * @see Native TEAL opcode: [`ec_add`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_add)
     * Min AVM version: 10
     */
    add(g, a, b) {
        throw new NoImplementation();
    },
    /**
     * maps field element A to group G
     * BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map.
     * G1 element inputs are base field elements and G2 element inputs are quadratic field elements, with nearly the same encoding rules (for field elements) as defined in `ec_add`. There is one difference of encoding rule: G1 element inputs do not need to be 0-padded if they fit in less than 32 bytes for BN254 and less than 48 bytes for BLS12-381. (As usual, the empty byte array represents 0.) G2 elements inputs need to be always have the required size.
     * @see Native TEAL opcode: [`ec_map_to`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_map_to)
     * Min AVM version: 10
     */
    mapTo(g, a) {
        throw new NoImplementation();
    },
    /**
     * for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn
     * A is a list of concatenated points, encoded and checked as described in `ec_add`. B is a list of concatenated scalars which, unlike ec_scalar_mul, must all be exactly 32 bytes long.
     * The name `ec_multi_scalar_mul` was chosen to reflect common usage, but a more consistent name would be `ec_multi_scalar_mul`. AVM values are limited to 4096 bytes, so `ec_multi_scalar_mul` is limited by the size of the points in the group being operated upon.
     * @see Native TEAL opcode: [`ec_multi_scalar_mul`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_multi_scalar_mul)
     * Min AVM version: 10
     */
    scalarMulMulti(g, a, b) {
        throw new NoImplementation();
    },
    /**
     * 1 if the product of the pairing of each point in A with its respective point in B is equal to the identity element of the target group Gt, else 0
     * A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes. AVM values are limited to 4096 bytes, so `ec_pairing_check` is limited by the size of the points in the groups being operated upon.
     * @see Native TEAL opcode: [`ec_pairing_check`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_pairing_check)
     * Min AVM version: 10
     */
    pairingCheck(g, a, b) {
        throw new NoImplementation();
    },
    /**
     * for curve point A and scalar B, return the curve point BA, the point A multiplied by the scalar B.
     * A is a curve point encoded and checked as described in `ec_add`. Scalar B is interpreted as a big-endian unsigned integer. Fails if B exceeds 32 bytes.
     * @see Native TEAL opcode: [`ec_scalar_mul`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_scalar_mul)
     * Min AVM version: 10
     */
    scalarMul(g, a, b) {
        throw new NoImplementation();
    },
    /**
     * 1 if A is in the main prime-order subgroup of G (including the point at infinity) else 0. Program fails if A is not in G at all.
     * @see Native TEAL opcode: [`ec_subgroup_check`](https://dev.algorand.co/reference/algorand-teal/opcodes#ec_subgroup_check)
     * Min AVM version: 10
     */
    subgroupCheck(g, a) {
        throw new NoImplementation();
    },
};
/**
 * decompress pubkey A into components X, Y
 * The 33 byte public key in a compressed form to be decompressed into X and Y (top) components. All values are big-endian encoded.
 * @see Native TEAL opcode: [`ecdsa_pk_decompress`](https://dev.algorand.co/reference/algorand-teal/opcodes#ecdsa_pk_decompress)
 * Min AVM version: 5
 */
function ecdsaPkDecompress(v, a) {
    throw new NoImplementation();
}
/**
 * for (data A, recovery id B, signature C, D) recover a public key
 * S (top) and R elements of a signature, recovery id and data (bottom) are expected on the stack and used to deriver a public key. All values are big-endian encoded. The signed data must be 32 bytes long.
 * @see Native TEAL opcode: [`ecdsa_pk_recover`](https://dev.algorand.co/reference/algorand-teal/opcodes#ecdsa_pk_recover)
 * Min AVM version: 5
 */
function ecdsaPkRecover(v, a, b, c, d) {
    throw new NoImplementation();
}
/**
 * for (data A, signature B, C and pubkey D, E) verify the signature of the data against the pubkey => {0 or 1}
 * The 32 byte Y-component of a public key is the last element on the stack, preceded by X-component of a pubkey, preceded by S and R components of a signature, preceded by the data that is fifth element on the stack. All values are big-endian encoded. The signed data must be 32 bytes long, and signatures in lower-S form are only accepted.
 * @see Native TEAL opcode: [`ecdsa_verify`](https://dev.algorand.co/reference/algorand-teal/opcodes#ecdsa_verify)
 * Min AVM version: 5
 */
function ecdsaVerify(v, a, b, c, d, e) {
    throw new NoImplementation();
}
/**
 * for (data A, signature B, pubkey C) verify the signature of ("ProgData" || program_hash || data) against the pubkey => {0 or 1}
 * The 32 byte public key is the last element on the stack, preceded by the 64 byte signature at the second-to-last element on the stack, preceded by the data which was signed at the third-to-last element on the stack.
 * @see Native TEAL opcode: [`ed25519verify`](https://dev.algorand.co/reference/algorand-teal/opcodes#ed25519verify)
 * Min AVM version: 1
 */
function ed25519verify(a, b, c) {
    throw new NoImplementation();
}
/**
 * for (data A, signature B, pubkey C) verify the signature of the data against the pubkey => {0 or 1}
 * @see Native TEAL opcode: [`ed25519verify_bare`](https://dev.algorand.co/reference/algorand-teal/opcodes#ed25519verify_bare)
 * Min AVM version: 7
 */
function ed25519verifyBare(a, b, c) {
    throw new NoImplementation();
}
/**
 * A raised to the Bth power. Fail if A == B == 0 and on overflow
 * @see Native TEAL opcode: [`exp`](https://dev.algorand.co/reference/algorand-teal/opcodes#exp)
 * Min AVM version: 4
 */
function exp(a, b) {
    throw new NoImplementation();
}
/**
 * A raised to the Bth power as a 128-bit result in two uint64s. X is the high 64 bits, Y is the low. Fail if A == B == 0 or if the results exceeds 2^128-1
 * @see Native TEAL opcode: [`expw`](https://dev.algorand.co/reference/algorand-teal/opcodes#expw)
 * Min AVM version: 4
 */
function expw(a, b) {
    throw new NoImplementation();
}
/**
 * A uint16 formed from a range of big-endian bytes from A starting at B up to but not including B+2. If B+2 is larger than the array length, the program fails
 * @see Native TEAL opcode: [`extract_uint16`](https://dev.algorand.co/reference/algorand-teal/opcodes#extract_uint16)
 * Min AVM version: 5
 */
function extractUint16(a, b) {
    throw new NoImplementation();
}
/**
 * A uint32 formed from a range of big-endian bytes from A starting at B up to but not including B+4. If B+4 is larger than the array length, the program fails
 * @see Native TEAL opcode: [`extract_uint32`](https://dev.algorand.co/reference/algorand-teal/opcodes#extract_uint32)
 * Min AVM version: 5
 */
function extractUint32(a, b) {
    throw new NoImplementation();
}
/**
 * A uint64 formed from a range of big-endian bytes from A starting at B up to but not including B+8. If B+8 is larger than the array length, the program fails
 * @see Native TEAL opcode: [`extract_uint64`](https://dev.algorand.co/reference/algorand-teal/opcodes#extract_uint64)
 * Min AVM version: 5
 */
function extractUint64(a, b) {
    throw new NoImplementation();
}
/**
 * for (data A, compressed-format signature B, pubkey C) verify the signature of data against the pubkey => {0 or 1}
 * @see Native TEAL opcode: [`falcon_verify`](https://dev.algorand.co/reference/algorand-teal/opcodes#falcon_verify)
 * Min AVM version: 12
 */
function falconVerify(a, b, c) {
    throw new NoImplementation();
}
/**
 * ID of the asset or application created in the Ath transaction of the current group
 * `gaids` fails unless the requested transaction created an asset or application and A < GroupIndex.
 * @see Native TEAL opcode: [`gaids`](https://dev.algorand.co/reference/algorand-teal/opcodes#gaids)
 * Min AVM version: 4
 */
function gaid(a) {
    throw new NoImplementation();
}
/**
 * Bth bit of (byte-array or integer) A. If B is greater than or equal to the bit length of the value (8*byte length), the program fails
 * see explanation of bit ordering in setbit
 * @see Native TEAL opcode: [`getbit`](https://dev.algorand.co/reference/algorand-teal/opcodes#getbit)
 * Min AVM version: 3
 */
function getBit(a, b) {
    throw new NoImplementation();
}
/**
 * Bth byte of A, as an integer. If B is greater than or equal to the array length, the program fails
 * @see Native TEAL opcode: [`getbyte`](https://dev.algorand.co/reference/algorand-teal/opcodes#getbyte)
 * Min AVM version: 3
 */
function getByte(a, b) {
    throw new NoImplementation();
}
/**
 * Get values for inner transaction in the last group submitted
 */
const GITxn = {
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    sender(t) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 6
     */
    fee(t) {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 6
     */
    firstValid(t) {
        throw new NoImplementation();
    },
    /**
     * UNIX timestamp of block before txn.FirstValid. Fails if negative
     * Min AVM version: 7
     */
    firstValidTime(t) {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 6
     */
    lastValid(t) {
        throw new NoImplementation();
    },
    /**
     * Any data up to 1024 bytes
     * Min AVM version: 6
     */
    note(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte lease value
     * Min AVM version: 6
     */
    lease(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    receiver(t) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 6
     */
    amount(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    closeRemainderTo(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    votePk(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    selectionPk(t) {
        throw new NoImplementation();
    },
    /**
     * The first round that the participation key is valid.
     * Min AVM version: 6
     */
    voteFirst(t) {
        throw new NoImplementation();
    },
    /**
     * The last round that the participation key is valid.
     * Min AVM version: 6
     */
    voteLast(t) {
        throw new NoImplementation();
    },
    /**
     * Dilution for the 2-level participation key
     * Min AVM version: 6
     */
    voteKeyDilution(t) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as bytes
     * Min AVM version: 6
     */
    type(t) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as integer
     * Min AVM version: 6
     */
    typeEnum(t) {
        throw new NoImplementation();
    },
    /**
     * Asset ID
     * Min AVM version: 6
     */
    xferAsset(t) {
        throw new NoImplementation();
    },
    /**
     * value in Asset's units
     * Min AVM version: 6
     */
    assetAmount(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address. Source of assets if Sender is the Asset's Clawback address.
     * Min AVM version: 6
     */
    assetSender(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    assetReceiver(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 6
     */
    assetCloseTo(t) {
        throw new NoImplementation();
    },
    /**
     * Position of this transaction within an atomic transaction group. A stand-alone transaction is implicitly element 0 in a group of 1
     * Min AVM version: 6
     */
    groupIndex(t) {
        throw new NoImplementation();
    },
    /**
     * The computed ID for this transaction. 32 bytes.
     * Min AVM version: 6
     */
    txId(t) {
        throw new NoImplementation();
    },
    /**
     * ApplicationID from ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationId(t) {
        throw new NoImplementation();
    },
    /**
     * ApplicationCall transaction on completion action
     * Min AVM version: 2
     */
    onCompletion(t) {
        throw new NoImplementation();
    },
    /**
     * Arguments passed to the application in the ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationArgs(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of ApplicationArgs
     * Min AVM version: 2
     */
    numAppArgs(t) {
        throw new NoImplementation();
    },
    /**
     * Accounts listed in the ApplicationCall transaction
     * Min AVM version: 2
     */
    accounts(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of Accounts
     * Min AVM version: 2
     */
    numAccounts(t) {
        throw new NoImplementation();
    },
    /**
     * Approval program
     * Min AVM version: 2
     */
    approvalProgram(t) {
        throw new NoImplementation();
    },
    /**
     * Clear state program
     * Min AVM version: 2
     */
    clearStateProgram(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte Sender's new AuthAddr
     * Min AVM version: 2
     */
    rekeyTo(t) {
        throw new NoImplementation();
    },
    /**
     * Asset ID in asset config transaction
     * Min AVM version: 2
     */
    configAsset(t) {
        throw new NoImplementation();
    },
    /**
     * Total number of units of this asset created
     * Min AVM version: 2
     */
    configAssetTotal(t) {
        throw new NoImplementation();
    },
    /**
     * Number of digits to display after the decimal place when displaying the asset
     * Min AVM version: 2
     */
    configAssetDecimals(t) {
        throw new NoImplementation();
    },
    /**
     * Whether the asset's slots are frozen by default or not, 0 or 1
     * Min AVM version: 2
     */
    configAssetDefaultFrozen(t) {
        throw new NoImplementation();
    },
    /**
     * Unit name of the asset
     * Min AVM version: 2
     */
    configAssetUnitName(t) {
        throw new NoImplementation();
    },
    /**
     * The asset name
     * Min AVM version: 2
     */
    configAssetName(t) {
        throw new NoImplementation();
    },
    /**
     * URL
     * Min AVM version: 2
     */
    configAssetUrl(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte commitment to unspecified asset metadata
     * Min AVM version: 2
     */
    configAssetMetadataHash(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetManager(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetReserve(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetFreeze(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetClawback(t) {
        throw new NoImplementation();
    },
    /**
     * Asset ID being frozen or un-frozen
     * Min AVM version: 2
     */
    freezeAsset(t) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of the account whose asset slot is being frozen or un-frozen
     * Min AVM version: 2
     */
    freezeAssetAccount(t) {
        throw new NoImplementation();
    },
    /**
     * The new frozen value, 0 or 1
     * Min AVM version: 2
     */
    freezeAssetFrozen(t) {
        throw new NoImplementation();
    },
    /**
     * Foreign Assets listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    assets(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of Assets
     * Min AVM version: 3
     */
    numAssets(t) {
        throw new NoImplementation();
    },
    /**
     * Foreign Apps listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    applications(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of Applications
     * Min AVM version: 3
     */
    numApplications(t) {
        throw new NoImplementation();
    },
    /**
     * Number of global state integers in ApplicationCall
     * Min AVM version: 3
     */
    globalNumUint(t) {
        throw new NoImplementation();
    },
    /**
     * Number of global state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    globalNumByteSlice(t) {
        throw new NoImplementation();
    },
    /**
     * Number of local state integers in ApplicationCall
     * Min AVM version: 3
     */
    localNumUint(t) {
        throw new NoImplementation();
    },
    /**
     * Number of local state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    localNumByteSlice(t) {
        throw new NoImplementation();
    },
    /**
     * Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program.
     * Min AVM version: 4
     */
    extraProgramPages(t) {
        throw new NoImplementation();
    },
    /**
     * Marks an account nonparticipating for rewards
     * Min AVM version: 5
     */
    nonparticipation(t) {
        throw new NoImplementation();
    },
    /**
     * Log messages emitted by an application call (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    logs(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of Logs (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    numLogs(t) {
        throw new NoImplementation();
    },
    /**
     * Asset ID allocated by the creation of an ASA (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    createdAssetId(t) {
        throw new NoImplementation();
    },
    /**
     * ApplicationID allocated by the creation of an application (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    createdApplicationId(t) {
        throw new NoImplementation();
    },
    /**
     * The last message emitted. Empty bytes if none were emitted. Application mode only
     * Min AVM version: 6
     */
    lastLog(t) {
        throw new NoImplementation();
    },
    /**
     * State proof public key
     * Min AVM version: 6
     */
    stateProofPk(t) {
        throw new NoImplementation();
    },
    /**
     * Approval Program as an array of pages
     * Min AVM version: 7
     */
    approvalProgramPages(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of Approval Program pages
     * Min AVM version: 7
     */
    numApprovalProgramPages(t) {
        throw new NoImplementation();
    },
    /**
     * ClearState Program as an array of pages
     * Min AVM version: 7
     */
    clearStateProgramPages(t, a) {
        throw new NoImplementation();
    },
    /**
     * Number of ClearState Program pages
     * Min AVM version: 7
     */
    numClearStateProgramPages(t) {
        throw new NoImplementation();
    },
    /**
     * Application version for which the txn must reject
     * Min AVM version: 12
     */
    rejectVersion(t) {
        throw new NoImplementation();
    },
};
/**
 * Bth scratch space value of the Ath transaction in the current group
 * @see Native TEAL opcode: [`gloadss`](https://dev.algorand.co/reference/algorand-teal/opcodes#gloadss)
 * Min AVM version: 6
 */
function gloadBytes(a, b) {
    throw new NoImplementation();
}
/**
 * Bth scratch space value of the Ath transaction in the current group
 * @see Native TEAL opcode: [`gloadss`](https://dev.algorand.co/reference/algorand-teal/opcodes#gloadss)
 * Min AVM version: 6
 */
function gloadUint64(a, b) {
    throw new NoImplementation();
}
const Global = {
    /**
     * microalgos
     * Min AVM version: 1
     */
    get minTxnFee() {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 1
     */
    get minBalance() {
        throw new NoImplementation();
    },
    /**
     * rounds
     * Min AVM version: 1
     */
    get maxTxnLife() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of all zero bytes
     * Min AVM version: 1
     */
    get zeroAddress() {
        throw new NoImplementation();
    },
    /**
     * Number of transactions in this atomic transaction group. At least 1
     * Min AVM version: 1
     */
    get groupSize() {
        throw new NoImplementation();
    },
    /**
     * Maximum supported version
     * Min AVM version: 2
     */
    get logicSigVersion() {
        throw new NoImplementation();
    },
    /**
     * Current round number. Application mode only.
     * Min AVM version: 2
     */
    get round() {
        throw new NoImplementation();
    },
    /**
     * Last confirmed block UNIX timestamp. Fails if negative. Application mode only.
     * Min AVM version: 2
     */
    get latestTimestamp() {
        throw new NoImplementation();
    },
    /**
     * ID of current application executing. Application mode only.
     * Min AVM version: 2
     */
    get currentApplicationId() {
        throw new NoImplementation();
    },
    /**
     * Address of the creator of the current application. Application mode only.
     * Min AVM version: 3
     */
    get creatorAddress() {
        throw new NoImplementation();
    },
    /**
     * Address that the current application controls. Application mode only.
     * Min AVM version: 5
     */
    get currentApplicationAddress() {
        throw new NoImplementation();
    },
    /**
     * ID of the transaction group. 32 zero bytes if the transaction is not part of a group.
     * Min AVM version: 5
     */
    get groupId() {
        throw new NoImplementation();
    },
    /**
     * The remaining cost that can be spent by opcodes in this program.
     * Min AVM version: 6
     */
    get opcodeBudget() {
        throw new NoImplementation();
    },
    /**
     * The application ID of the application that called this application. 0 if this application is at the top-level. Application mode only.
     * Min AVM version: 6
     */
    get callerApplicationId() {
        throw new NoImplementation();
    },
    /**
     * The application address of the application that called this application. ZeroAddress if this application is at the top-level. Application mode only.
     * Min AVM version: 6
     */
    get callerApplicationAddress() {
        throw new NoImplementation();
    },
    /**
     * The additional minimum balance required to create (and opt-in to) an asset.
     * Min AVM version: 10
     */
    get assetCreateMinBalance() {
        throw new NoImplementation();
    },
    /**
     * The additional minimum balance required to opt-in to an asset.
     * Min AVM version: 10
     */
    get assetOptInMinBalance() {
        throw new NoImplementation();
    },
    /**
     * The Genesis Hash for the network.
     * Min AVM version: 10
     */
    get genesisHash() {
        throw new NoImplementation();
    },
    /**
     * Whether block proposal payouts are enabled.
     * Min AVM version: 11
     */
    get payoutsEnabled() {
        throw new NoImplementation();
    },
    /**
     * The fee required in a keyreg transaction to make an account incentive eligible.
     * Min AVM version: 11
     */
    get payoutsGoOnlineFee() {
        throw new NoImplementation();
    },
    /**
     * The percentage of transaction fees in a block that can be paid to the block proposer.
     * Min AVM version: 11
     */
    get payoutsPercent() {
        throw new NoImplementation();
    },
    /**
     * The minimum balance an account must have in the agreement round to receive block payouts in the proposal round.
     * Min AVM version: 11
     */
    get payoutsMinBalance() {
        throw new NoImplementation();
    },
    /**
     * The maximum balance an account can have in the agreement round to receive block payouts in the proposal round.
     * Min AVM version: 11
     */
    get payoutsMaxBalance() {
        throw new NoImplementation();
    },
};
/**
 * Get values for transactions in the current group
 */
const GTxn = {
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    sender(a) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 3
     */
    fee(a) {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 3
     */
    firstValid(a) {
        throw new NoImplementation();
    },
    /**
     * UNIX timestamp of block before txn.FirstValid. Fails if negative
     * Min AVM version: 7
     */
    firstValidTime(a) {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 3
     */
    lastValid(a) {
        throw new NoImplementation();
    },
    /**
     * Any data up to 1024 bytes
     * Min AVM version: 3
     */
    note(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte lease value
     * Min AVM version: 3
     */
    lease(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    receiver(a) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 3
     */
    amount(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    closeRemainderTo(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    votePk(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    selectionPk(a) {
        throw new NoImplementation();
    },
    /**
     * The first round that the participation key is valid.
     * Min AVM version: 3
     */
    voteFirst(a) {
        throw new NoImplementation();
    },
    /**
     * The last round that the participation key is valid.
     * Min AVM version: 3
     */
    voteLast(a) {
        throw new NoImplementation();
    },
    /**
     * Dilution for the 2-level participation key
     * Min AVM version: 3
     */
    voteKeyDilution(a) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as bytes
     * Min AVM version: 3
     */
    type(a) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as integer
     * Min AVM version: 3
     */
    typeEnum(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID
     * Min AVM version: 3
     */
    xferAsset(a) {
        throw new NoImplementation();
    },
    /**
     * value in Asset's units
     * Min AVM version: 3
     */
    assetAmount(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address. Source of assets if Sender is the Asset's Clawback address.
     * Min AVM version: 3
     */
    assetSender(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    assetReceiver(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 3
     */
    assetCloseTo(a) {
        throw new NoImplementation();
    },
    /**
     * Position of this transaction within an atomic transaction group. A stand-alone transaction is implicitly element 0 in a group of 1
     * Min AVM version: 3
     */
    groupIndex(a) {
        throw new NoImplementation();
    },
    /**
     * The computed ID for this transaction. 32 bytes.
     * Min AVM version: 3
     */
    txId(a) {
        throw new NoImplementation();
    },
    /**
     * ApplicationID from ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationId(a) {
        throw new NoImplementation();
    },
    /**
     * ApplicationCall transaction on completion action
     * Min AVM version: 2
     */
    onCompletion(a) {
        throw new NoImplementation();
    },
    /**
     * Arguments passed to the application in the ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationArgs(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of ApplicationArgs
     * Min AVM version: 2
     */
    numAppArgs(a) {
        throw new NoImplementation();
    },
    /**
     * Accounts listed in the ApplicationCall transaction
     * Min AVM version: 2
     */
    accounts(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of Accounts
     * Min AVM version: 2
     */
    numAccounts(a) {
        throw new NoImplementation();
    },
    /**
     * Approval program
     * Min AVM version: 2
     */
    approvalProgram(a) {
        throw new NoImplementation();
    },
    /**
     * Clear state program
     * Min AVM version: 2
     */
    clearStateProgram(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte Sender's new AuthAddr
     * Min AVM version: 2
     */
    rekeyTo(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID in asset config transaction
     * Min AVM version: 2
     */
    configAsset(a) {
        throw new NoImplementation();
    },
    /**
     * Total number of units of this asset created
     * Min AVM version: 2
     */
    configAssetTotal(a) {
        throw new NoImplementation();
    },
    /**
     * Number of digits to display after the decimal place when displaying the asset
     * Min AVM version: 2
     */
    configAssetDecimals(a) {
        throw new NoImplementation();
    },
    /**
     * Whether the asset's slots are frozen by default or not, 0 or 1
     * Min AVM version: 2
     */
    configAssetDefaultFrozen(a) {
        throw new NoImplementation();
    },
    /**
     * Unit name of the asset
     * Min AVM version: 2
     */
    configAssetUnitName(a) {
        throw new NoImplementation();
    },
    /**
     * The asset name
     * Min AVM version: 2
     */
    configAssetName(a) {
        throw new NoImplementation();
    },
    /**
     * URL
     * Min AVM version: 2
     */
    configAssetUrl(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte commitment to unspecified asset metadata
     * Min AVM version: 2
     */
    configAssetMetadataHash(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetManager(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetReserve(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetFreeze(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    configAssetClawback(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID being frozen or un-frozen
     * Min AVM version: 2
     */
    freezeAsset(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of the account whose asset slot is being frozen or un-frozen
     * Min AVM version: 2
     */
    freezeAssetAccount(a) {
        throw new NoImplementation();
    },
    /**
     * The new frozen value, 0 or 1
     * Min AVM version: 2
     */
    freezeAssetFrozen(a) {
        throw new NoImplementation();
    },
    /**
     * Foreign Assets listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    assets(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of Assets
     * Min AVM version: 3
     */
    numAssets(a) {
        throw new NoImplementation();
    },
    /**
     * Foreign Apps listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    applications(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of Applications
     * Min AVM version: 3
     */
    numApplications(a) {
        throw new NoImplementation();
    },
    /**
     * Number of global state integers in ApplicationCall
     * Min AVM version: 3
     */
    globalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of global state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    globalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of local state integers in ApplicationCall
     * Min AVM version: 3
     */
    localNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of local state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    localNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program.
     * Min AVM version: 4
     */
    extraProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Marks an account nonparticipating for rewards
     * Min AVM version: 5
     */
    nonparticipation(a) {
        throw new NoImplementation();
    },
    /**
     * Log messages emitted by an application call (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    logs(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of Logs (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    numLogs(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID allocated by the creation of an ASA (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    createdAssetId(a) {
        throw new NoImplementation();
    },
    /**
     * ApplicationID allocated by the creation of an application (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    createdApplicationId(a) {
        throw new NoImplementation();
    },
    /**
     * The last message emitted. Empty bytes if none were emitted. Application mode only
     * Min AVM version: 6
     */
    lastLog(a) {
        throw new NoImplementation();
    },
    /**
     * State proof public key
     * Min AVM version: 6
     */
    stateProofPk(a) {
        throw new NoImplementation();
    },
    /**
     * Approval Program as an array of pages
     * Min AVM version: 7
     */
    approvalProgramPages(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of Approval Program pages
     * Min AVM version: 7
     */
    numApprovalProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * ClearState Program as an array of pages
     * Min AVM version: 7
     */
    clearStateProgramPages(a, b) {
        throw new NoImplementation();
    },
    /**
     * Number of ClearState Program pages
     * Min AVM version: 7
     */
    numClearStateProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Application version for which the txn must reject
     * Min AVM version: 12
     */
    rejectVersion(a) {
        throw new NoImplementation();
    },
};
/**
 * converts uint64 A to big-endian byte array, always of length 8
 * @see Native TEAL opcode: [`itob`](https://dev.algorand.co/reference/algorand-teal/opcodes#itob)
 * Min AVM version: 1
 */
function itob(a) {
    throw new NoImplementation();
}
/**
 * Get values for the last inner transaction
 */
const ITxn = {
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get sender() {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 5
     */
    get fee() {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 5
     */
    get firstValid() {
        throw new NoImplementation();
    },
    /**
     * UNIX timestamp of block before txn.FirstValid. Fails if negative
     * Min AVM version: 7
     */
    get firstValidTime() {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 5
     */
    get lastValid() {
        throw new NoImplementation();
    },
    /**
     * Any data up to 1024 bytes
     * Min AVM version: 5
     */
    get note() {
        throw new NoImplementation();
    },
    /**
     * 32 byte lease value
     * Min AVM version: 5
     */
    get lease() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get receiver() {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 5
     */
    get amount() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get closeRemainderTo() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get votePk() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get selectionPk() {
        throw new NoImplementation();
    },
    /**
     * The first round that the participation key is valid.
     * Min AVM version: 5
     */
    get voteFirst() {
        throw new NoImplementation();
    },
    /**
     * The last round that the participation key is valid.
     * Min AVM version: 5
     */
    get voteLast() {
        throw new NoImplementation();
    },
    /**
     * Dilution for the 2-level participation key
     * Min AVM version: 5
     */
    get voteKeyDilution() {
        throw new NoImplementation();
    },
    /**
     * Transaction type as bytes
     * Min AVM version: 5
     */
    get type() {
        throw new NoImplementation();
    },
    /**
     * Transaction type as integer
     * Min AVM version: 5
     */
    get typeEnum() {
        throw new NoImplementation();
    },
    /**
     * Asset ID
     * Min AVM version: 5
     */
    get xferAsset() {
        throw new NoImplementation();
    },
    /**
     * value in Asset's units
     * Min AVM version: 5
     */
    get assetAmount() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address. Source of assets if Sender is the Asset's Clawback address.
     * Min AVM version: 5
     */
    get assetSender() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get assetReceiver() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    get assetCloseTo() {
        throw new NoImplementation();
    },
    /**
     * Position of this transaction within an atomic transaction group. A stand-alone transaction is implicitly element 0 in a group of 1
     * Min AVM version: 5
     */
    get groupIndex() {
        throw new NoImplementation();
    },
    /**
     * The computed ID for this transaction. 32 bytes.
     * Min AVM version: 5
     */
    get txId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationID from ApplicationCall transaction
     * Min AVM version: 2
     */
    get applicationId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationCall transaction on completion action
     * Min AVM version: 2
     */
    get onCompletion() {
        throw new NoImplementation();
    },
    /**
     * Arguments passed to the application in the ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationArgs(a) {
        throw new NoImplementation();
    },
    /**
     * Number of ApplicationArgs
     * Min AVM version: 2
     */
    get numAppArgs() {
        throw new NoImplementation();
    },
    /**
     * Accounts listed in the ApplicationCall transaction
     * Min AVM version: 2
     */
    accounts(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Accounts
     * Min AVM version: 2
     */
    get numAccounts() {
        throw new NoImplementation();
    },
    /**
     * Approval program
     * Min AVM version: 2
     */
    get approvalProgram() {
        throw new NoImplementation();
    },
    /**
     * Clear state program
     * Min AVM version: 2
     */
    get clearStateProgram() {
        throw new NoImplementation();
    },
    /**
     * 32 byte Sender's new AuthAddr
     * Min AVM version: 2
     */
    get rekeyTo() {
        throw new NoImplementation();
    },
    /**
     * Asset ID in asset config transaction
     * Min AVM version: 2
     */
    get configAsset() {
        throw new NoImplementation();
    },
    /**
     * Total number of units of this asset created
     * Min AVM version: 2
     */
    get configAssetTotal() {
        throw new NoImplementation();
    },
    /**
     * Number of digits to display after the decimal place when displaying the asset
     * Min AVM version: 2
     */
    get configAssetDecimals() {
        throw new NoImplementation();
    },
    /**
     * Whether the asset's slots are frozen by default or not, 0 or 1
     * Min AVM version: 2
     */
    get configAssetDefaultFrozen() {
        throw new NoImplementation();
    },
    /**
     * Unit name of the asset
     * Min AVM version: 2
     */
    get configAssetUnitName() {
        throw new NoImplementation();
    },
    /**
     * The asset name
     * Min AVM version: 2
     */
    get configAssetName() {
        throw new NoImplementation();
    },
    /**
     * URL
     * Min AVM version: 2
     */
    get configAssetUrl() {
        throw new NoImplementation();
    },
    /**
     * 32 byte commitment to unspecified asset metadata
     * Min AVM version: 2
     */
    get configAssetMetadataHash() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetManager() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetReserve() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetFreeze() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetClawback() {
        throw new NoImplementation();
    },
    /**
     * Asset ID being frozen or un-frozen
     * Min AVM version: 2
     */
    get freezeAsset() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of the account whose asset slot is being frozen or un-frozen
     * Min AVM version: 2
     */
    get freezeAssetAccount() {
        throw new NoImplementation();
    },
    /**
     * The new frozen value, 0 or 1
     * Min AVM version: 2
     */
    get freezeAssetFrozen() {
        throw new NoImplementation();
    },
    /**
     * Foreign Assets listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    assets(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Assets
     * Min AVM version: 3
     */
    get numAssets() {
        throw new NoImplementation();
    },
    /**
     * Foreign Apps listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    applications(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Applications
     * Min AVM version: 3
     */
    get numApplications() {
        throw new NoImplementation();
    },
    /**
     * Number of global state integers in ApplicationCall
     * Min AVM version: 3
     */
    get globalNumUint() {
        throw new NoImplementation();
    },
    /**
     * Number of global state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    get globalNumByteSlice() {
        throw new NoImplementation();
    },
    /**
     * Number of local state integers in ApplicationCall
     * Min AVM version: 3
     */
    get localNumUint() {
        throw new NoImplementation();
    },
    /**
     * Number of local state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    get localNumByteSlice() {
        throw new NoImplementation();
    },
    /**
     * Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program.
     * Min AVM version: 4
     */
    get extraProgramPages() {
        throw new NoImplementation();
    },
    /**
     * Marks an account nonparticipating for rewards
     * Min AVM version: 5
     */
    get nonparticipation() {
        throw new NoImplementation();
    },
    /**
     * Log messages emitted by an application call (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    logs(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Logs (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get numLogs() {
        throw new NoImplementation();
    },
    /**
     * Asset ID allocated by the creation of an ASA (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get createdAssetId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationID allocated by the creation of an application (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get createdApplicationId() {
        throw new NoImplementation();
    },
    /**
     * The last message emitted. Empty bytes if none were emitted. Application mode only
     * Min AVM version: 6
     */
    get lastLog() {
        throw new NoImplementation();
    },
    /**
     * State proof public key
     * Min AVM version: 6
     */
    get stateProofPk() {
        throw new NoImplementation();
    },
    /**
     * Approval Program as an array of pages
     * Min AVM version: 7
     */
    approvalProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Approval Program pages
     * Min AVM version: 7
     */
    get numApprovalProgramPages() {
        throw new NoImplementation();
    },
    /**
     * ClearState Program as an array of pages
     * Min AVM version: 7
     */
    clearStateProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Number of ClearState Program pages
     * Min AVM version: 7
     */
    get numClearStateProgramPages() {
        throw new NoImplementation();
    },
    /**
     * Application version for which the txn must reject
     * Min AVM version: 12
     */
    get rejectVersion() {
        throw new NoImplementation();
    },
};
/**
 * Create inner transactions
 */
const ITxnCreate = {
    /**
     * begin preparation of a new inner transaction in a new transaction group
     * `itxn_begin` initializes Sender to the application address; Fee to the minimum allowable, taking into account MinTxnFee and credit from overpaying in earlier transactions; FirstValid/LastValid to the values in the invoking transaction, and all other fields to zero or empty values.
     * @see Native TEAL opcode: [`itxn_begin`](https://dev.algorand.co/reference/algorand-teal/opcodes#itxn_begin)
     * Min AVM version: 5
     */
    begin() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setSender(a) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 5
     */
    setFee(a) {
        throw new NoImplementation();
    },
    /**
     * Any data up to 1024 bytes
     * Min AVM version: 5
     */
    setNote(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setReceiver(a) {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 5
     */
    setAmount(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setCloseRemainderTo(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setVotePk(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setSelectionPk(a) {
        throw new NoImplementation();
    },
    /**
     * The first round that the participation key is valid.
     * Min AVM version: 5
     */
    setVoteFirst(a) {
        throw new NoImplementation();
    },
    /**
     * The last round that the participation key is valid.
     * Min AVM version: 5
     */
    setVoteLast(a) {
        throw new NoImplementation();
    },
    /**
     * Dilution for the 2-level participation key
     * Min AVM version: 5
     */
    setVoteKeyDilution(a) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as bytes
     * Min AVM version: 5
     */
    setType(a) {
        throw new NoImplementation();
    },
    /**
     * Transaction type as integer
     * Min AVM version: 5
     */
    setTypeEnum(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID
     * Min AVM version: 5
     */
    setXferAsset(a) {
        throw new NoImplementation();
    },
    /**
     * value in Asset's units
     * Min AVM version: 5
     */
    setAssetAmount(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address. Source of assets if Sender is the Asset's Clawback address.
     * Min AVM version: 5
     */
    setAssetSender(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setAssetReceiver(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 5
     */
    setAssetCloseTo(a) {
        throw new NoImplementation();
    },
    /**
     * ApplicationID from ApplicationCall transaction
     * Min AVM version: 2
     */
    setApplicationId(a) {
        throw new NoImplementation();
    },
    /**
     * ApplicationCall transaction on completion action
     * Min AVM version: 2
     */
    setOnCompletion(a) {
        throw new NoImplementation();
    },
    /**
     * Arguments passed to the application in the ApplicationCall transaction
     * Min AVM version: 2
     */
    setApplicationArgs(a) {
        throw new NoImplementation();
    },
    /**
     * Accounts listed in the ApplicationCall transaction
     * Min AVM version: 2
     */
    setAccounts(a) {
        throw new NoImplementation();
    },
    /**
     * Approval program
     * Min AVM version: 2
     */
    setApprovalProgram(a) {
        throw new NoImplementation();
    },
    /**
     * Clear state program
     * Min AVM version: 2
     */
    setClearStateProgram(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte Sender's new AuthAddr
     * Min AVM version: 2
     */
    setRekeyTo(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID in asset config transaction
     * Min AVM version: 2
     */
    setConfigAsset(a) {
        throw new NoImplementation();
    },
    /**
     * Total number of units of this asset created
     * Min AVM version: 2
     */
    setConfigAssetTotal(a) {
        throw new NoImplementation();
    },
    /**
     * Number of digits to display after the decimal place when displaying the asset
     * Min AVM version: 2
     */
    setConfigAssetDecimals(a) {
        throw new NoImplementation();
    },
    /**
     * Whether the asset's slots are frozen by default or not, 0 or 1
     * Min AVM version: 2
     */
    setConfigAssetDefaultFrozen(a) {
        throw new NoImplementation();
    },
    /**
     * Unit name of the asset
     * Min AVM version: 2
     */
    setConfigAssetUnitName(a) {
        throw new NoImplementation();
    },
    /**
     * The asset name
     * Min AVM version: 2
     */
    setConfigAssetName(a) {
        throw new NoImplementation();
    },
    /**
     * URL
     * Min AVM version: 2
     */
    setConfigAssetUrl(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte commitment to unspecified asset metadata
     * Min AVM version: 2
     */
    setConfigAssetMetadataHash(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    setConfigAssetManager(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    setConfigAssetReserve(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    setConfigAssetFreeze(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    setConfigAssetClawback(a) {
        throw new NoImplementation();
    },
    /**
     * Asset ID being frozen or un-frozen
     * Min AVM version: 2
     */
    setFreezeAsset(a) {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of the account whose asset slot is being frozen or un-frozen
     * Min AVM version: 2
     */
    setFreezeAssetAccount(a) {
        throw new NoImplementation();
    },
    /**
     * The new frozen value, 0 or 1
     * Min AVM version: 2
     */
    setFreezeAssetFrozen(a) {
        throw new NoImplementation();
    },
    /**
     * Foreign Assets listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    setAssets(a) {
        throw new NoImplementation();
    },
    /**
     * Foreign Apps listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    setApplications(a) {
        throw new NoImplementation();
    },
    /**
     * Number of global state integers in ApplicationCall
     * Min AVM version: 3
     */
    setGlobalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of global state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    setGlobalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of local state integers in ApplicationCall
     * Min AVM version: 3
     */
    setLocalNumUint(a) {
        throw new NoImplementation();
    },
    /**
     * Number of local state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    setLocalNumByteSlice(a) {
        throw new NoImplementation();
    },
    /**
     * Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program.
     * Min AVM version: 4
     */
    setExtraProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Marks an account nonparticipating for rewards
     * Min AVM version: 5
     */
    setNonparticipation(a) {
        throw new NoImplementation();
    },
    /**
     * State proof public key
     * Min AVM version: 6
     */
    setStateProofPk(a) {
        throw new NoImplementation();
    },
    /**
     * Approval Program as an array of pages
     * Min AVM version: 7
     */
    setApprovalProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * ClearState Program as an array of pages
     * Min AVM version: 7
     */
    setClearStateProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Application version for which the txn must reject
     * Min AVM version: 12
     */
    setRejectVersion(a) {
        throw new NoImplementation();
    },
    /**
     * begin preparation of a new inner transaction in the same transaction group
     * `itxn_next` initializes the transaction exactly as `itxn_begin` does
     * @see Native TEAL opcode: [`itxn_next`](https://dev.algorand.co/reference/algorand-teal/opcodes#itxn_next)
     * Min AVM version: 6
     */
    next() {
        throw new NoImplementation();
    },
    /**
     * execute the current inner transaction group. Fail if executing this group would exceed the inner transaction limit, or if any transaction in the group fails.
     * `itxn_submit` resets the current transaction so that it can not be resubmitted. A new `itxn_begin` is required to prepare another inner transaction.
     * @see Native TEAL opcode: [`itxn_submit`](https://dev.algorand.co/reference/algorand-teal/opcodes#itxn_submit)
     * Min AVM version: 5
     */
    submit() {
        throw new NoImplementation();
    },
};
const JsonRef = {
    jsonString(a, b) {
        throw new NoImplementation();
    },
    jsonUint64(a, b) {
        throw new NoImplementation();
    },
    jsonObject(a, b) {
        throw new NoImplementation();
    },
};
/**
 * Keccak256 hash of value A, yields [32]byte
 * @see Native TEAL opcode: [`keccak256`](https://dev.algorand.co/reference/algorand-teal/opcodes#keccak256)
 * Min AVM version: 1
 */
function keccak256(a) {
    throw new NoImplementation();
}
/**
 * yields length of byte value A
 * @see Native TEAL opcode: [`len`](https://dev.algorand.co/reference/algorand-teal/opcodes#len)
 * Min AVM version: 1
 */
function len(a) {
    throw new NoImplementation();
}
/**
 * Load or store scratch values
 */
const Scratch = {
    /**
     * Ath scratch space value.  All scratch spaces are 0 at program start.
     * @see Native TEAL opcode: [`loads`](https://dev.algorand.co/reference/algorand-teal/opcodes#loads)
     * Min AVM version: 5
     */
    loadBytes(a) {
        throw new NoImplementation();
    },
    /**
     * Ath scratch space value.  All scratch spaces are 0 at program start.
     * @see Native TEAL opcode: [`loads`](https://dev.algorand.co/reference/algorand-teal/opcodes#loads)
     * Min AVM version: 5
     */
    loadUint64(a) {
        throw new NoImplementation();
    },
    /**
     * store B to the Ath scratch space
     * @see Native TEAL opcode: [`stores`](https://dev.algorand.co/reference/algorand-teal/opcodes#stores)
     * Min AVM version: 5
     */
    store(a, b) {
        throw new NoImplementation();
    },
};
/**
 * MiMC hash of scalars A, using curve and parameters specified by configuration C
 * A is a list of concatenated 32 byte big-endian unsigned integer scalars.  Fail if A's length is not a multiple of 32 or any element exceeds the curve modulus.
 * The MiMC hash function has known collisions since any input which is a multiple of the elliptic curve modulus will hash to the same value. MiMC is thus not a general purpose hash function, but meant to be used in zero knowledge applications to match a zk-circuit implementation.
 * @see Native TEAL opcode: [`mimc`](https://dev.algorand.co/reference/algorand-teal/opcodes#mimc)
 * Min AVM version: 11
 */
function mimc(c, a) {
    throw new NoImplementation();
}
/**
 * minimum required balance for account A, in microalgos. Required balance is affected by ASA, App, and Box usage. When creating or opting into an app, the minimum balance grows before the app code runs, therefore the increase is visible there. When deleting or closing out, the minimum balance decreases after the app executes. Changes caused by inner transactions or box usage are observable immediately following the opcode effecting the change.
 * @param a Txn.Accounts offset (or, since v4, an _available_ account address), _available_ application id (or, since v4, a Txn.ForeignApps offset).
 * @return value.
 * @see Native TEAL opcode: [`min_balance`](https://dev.algorand.co/reference/algorand-teal/opcodes#min_balance)
 * Min AVM version: 3
 */
function minBalance(a) {
    throw new NoImplementation();
}
/**
 * A times B as a 128-bit result in two uint64s. X is the high 64 bits, Y is the low
 * @see Native TEAL opcode: [`mulw`](https://dev.algorand.co/reference/algorand-teal/opcodes#mulw)
 * Min AVM version: 1
 */
function mulw(a, b) {
    throw new NoImplementation();
}
/**
 * the total online stake in the agreement round
 * @see Native TEAL opcode: [`online_stake`](https://dev.algorand.co/reference/algorand-teal/opcodes#online_stake)
 * Min AVM version: 11
 */
function onlineStake() {
    throw new NoImplementation();
}
/**
 * Copy of A with the bytes starting at B replaced by the bytes of C. Fails if B+len(C) exceeds len(A)
 * `replace3` can be called using `replace` with no immediates.
 * @see Native TEAL opcode: [`replace3`](https://dev.algorand.co/reference/algorand-teal/opcodes#replace3)
 * Min AVM version: 7
 */
function replace(a, b, c) {
    throw new NoImplementation();
}
/**
 * Copy of A with the Bth byte set to small integer (between 0..255) C. If B is greater than or equal to the array length, the program fails
 * @see Native TEAL opcode: [`setbyte`](https://dev.algorand.co/reference/algorand-teal/opcodes#setbyte)
 * Min AVM version: 3
 */
function setByte(a, b, c) {
    throw new NoImplementation();
}
/**
 * SHA256 hash of value A, yields [32]byte
 * @see Native TEAL opcode: [`sha256`](https://dev.algorand.co/reference/algorand-teal/opcodes#sha256)
 * Min AVM version: 1
 */
function sha256(a) {
    throw new NoImplementation();
}
/**
 * SHA3_256 hash of value A, yields [32]byte
 * @see Native TEAL opcode: [`sha3_256`](https://dev.algorand.co/reference/algorand-teal/opcodes#sha3_256)
 * Min AVM version: 7
 */
function sha3_256(a) {
    throw new NoImplementation();
}
/**
 * SHA512_256 hash of value A, yields [32]byte
 * @see Native TEAL opcode: [`sha512_256`](https://dev.algorand.co/reference/algorand-teal/opcodes#sha512_256)
 * Min AVM version: 1
 */
function sha512_256(a) {
    throw new NoImplementation();
}
/**
 * A times 2^B, modulo 2^64
 * @see Native TEAL opcode: [`shl`](https://dev.algorand.co/reference/algorand-teal/opcodes#shl)
 * Min AVM version: 4
 */
function shl(a, b) {
    throw new NoImplementation();
}
/**
 * A divided by 2^B
 * @see Native TEAL opcode: [`shr`](https://dev.algorand.co/reference/algorand-teal/opcodes#shr)
 * Min AVM version: 4
 */
function shr(a, b) {
    throw new NoImplementation();
}
/**
 * The largest integer I such that I^2 <= A
 * @see Native TEAL opcode: [`sqrt`](https://dev.algorand.co/reference/algorand-teal/opcodes#sqrt)
 * Min AVM version: 4
 */
function sqrt(a) {
    throw new NoImplementation();
}
/**
 * A range of bytes from A starting at B up to but not including C. If C < B, or either is larger than the array length, the program fails
 * @see Native TEAL opcode: [`substring3`](https://dev.algorand.co/reference/algorand-teal/opcodes#substring3)
 * Min AVM version: 2
 */
function substring(a, b, c) {
    throw new NoImplementation();
}
/**
 * sumhash512 of value A, yields [64]byte
 * @see Native TEAL opcode: [`sumhash512`](https://dev.algorand.co/reference/algorand-teal/opcodes#sumhash512)
 * Min AVM version: 13
 */
function sumhash512(a) {
    throw new NoImplementation();
}
/**
 * Get values for the current executing transaction
 */
const Txn = {
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get sender() {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 1
     */
    get fee() {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 1
     */
    get firstValid() {
        throw new NoImplementation();
    },
    /**
     * UNIX timestamp of block before txn.FirstValid. Fails if negative
     * Min AVM version: 7
     */
    get firstValidTime() {
        throw new NoImplementation();
    },
    /**
     * round number
     * Min AVM version: 1
     */
    get lastValid() {
        throw new NoImplementation();
    },
    /**
     * Any data up to 1024 bytes
     * Min AVM version: 1
     */
    get note() {
        throw new NoImplementation();
    },
    /**
     * 32 byte lease value
     * Min AVM version: 1
     */
    get lease() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get receiver() {
        throw new NoImplementation();
    },
    /**
     * microalgos
     * Min AVM version: 1
     */
    get amount() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get closeRemainderTo() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get votePk() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get selectionPk() {
        throw new NoImplementation();
    },
    /**
     * The first round that the participation key is valid.
     * Min AVM version: 1
     */
    get voteFirst() {
        throw new NoImplementation();
    },
    /**
     * The last round that the participation key is valid.
     * Min AVM version: 1
     */
    get voteLast() {
        throw new NoImplementation();
    },
    /**
     * Dilution for the 2-level participation key
     * Min AVM version: 1
     */
    get voteKeyDilution() {
        throw new NoImplementation();
    },
    /**
     * Transaction type as bytes
     * Min AVM version: 1
     */
    get type() {
        throw new NoImplementation();
    },
    /**
     * Transaction type as integer
     * Min AVM version: 1
     */
    get typeEnum() {
        throw new NoImplementation();
    },
    /**
     * Asset ID
     * Min AVM version: 1
     */
    get xferAsset() {
        throw new NoImplementation();
    },
    /**
     * value in Asset's units
     * Min AVM version: 1
     */
    get assetAmount() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address. Source of assets if Sender is the Asset's Clawback address.
     * Min AVM version: 1
     */
    get assetSender() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get assetReceiver() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 1
     */
    get assetCloseTo() {
        throw new NoImplementation();
    },
    /**
     * Position of this transaction within an atomic transaction group. A stand-alone transaction is implicitly element 0 in a group of 1
     * Min AVM version: 1
     */
    get groupIndex() {
        throw new NoImplementation();
    },
    /**
     * The computed ID for this transaction. 32 bytes.
     * Min AVM version: 1
     */
    get txId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationID from ApplicationCall transaction
     * Min AVM version: 2
     */
    get applicationId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationCall transaction on completion action
     * Min AVM version: 2
     */
    get onCompletion() {
        throw new NoImplementation();
    },
    /**
     * Arguments passed to the application in the ApplicationCall transaction
     * Min AVM version: 2
     */
    applicationArgs(a) {
        throw new NoImplementation();
    },
    /**
     * Number of ApplicationArgs
     * Min AVM version: 2
     */
    get numAppArgs() {
        throw new NoImplementation();
    },
    /**
     * Accounts listed in the ApplicationCall transaction
     * Min AVM version: 2
     */
    accounts(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Accounts
     * Min AVM version: 2
     */
    get numAccounts() {
        throw new NoImplementation();
    },
    /**
     * Approval program
     * Min AVM version: 2
     */
    get approvalProgram() {
        throw new NoImplementation();
    },
    /**
     * Clear state program
     * Min AVM version: 2
     */
    get clearStateProgram() {
        throw new NoImplementation();
    },
    /**
     * 32 byte Sender's new AuthAddr
     * Min AVM version: 2
     */
    get rekeyTo() {
        throw new NoImplementation();
    },
    /**
     * Asset ID in asset config transaction
     * Min AVM version: 2
     */
    get configAsset() {
        throw new NoImplementation();
    },
    /**
     * Total number of units of this asset created
     * Min AVM version: 2
     */
    get configAssetTotal() {
        throw new NoImplementation();
    },
    /**
     * Number of digits to display after the decimal place when displaying the asset
     * Min AVM version: 2
     */
    get configAssetDecimals() {
        throw new NoImplementation();
    },
    /**
     * Whether the asset's slots are frozen by default or not, 0 or 1
     * Min AVM version: 2
     */
    get configAssetDefaultFrozen() {
        throw new NoImplementation();
    },
    /**
     * Unit name of the asset
     * Min AVM version: 2
     */
    get configAssetUnitName() {
        throw new NoImplementation();
    },
    /**
     * The asset name
     * Min AVM version: 2
     */
    get configAssetName() {
        throw new NoImplementation();
    },
    /**
     * URL
     * Min AVM version: 2
     */
    get configAssetUrl() {
        throw new NoImplementation();
    },
    /**
     * 32 byte commitment to unspecified asset metadata
     * Min AVM version: 2
     */
    get configAssetMetadataHash() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetManager() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetReserve() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetFreeze() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address
     * Min AVM version: 2
     */
    get configAssetClawback() {
        throw new NoImplementation();
    },
    /**
     * Asset ID being frozen or un-frozen
     * Min AVM version: 2
     */
    get freezeAsset() {
        throw new NoImplementation();
    },
    /**
     * 32 byte address of the account whose asset slot is being frozen or un-frozen
     * Min AVM version: 2
     */
    get freezeAssetAccount() {
        throw new NoImplementation();
    },
    /**
     * The new frozen value, 0 or 1
     * Min AVM version: 2
     */
    get freezeAssetFrozen() {
        throw new NoImplementation();
    },
    /**
     * Foreign Assets listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    assets(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Assets
     * Min AVM version: 3
     */
    get numAssets() {
        throw new NoImplementation();
    },
    /**
     * Foreign Apps listed in the ApplicationCall transaction
     * Min AVM version: 3
     */
    applications(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Applications
     * Min AVM version: 3
     */
    get numApplications() {
        throw new NoImplementation();
    },
    /**
     * Number of global state integers in ApplicationCall
     * Min AVM version: 3
     */
    get globalNumUint() {
        throw new NoImplementation();
    },
    /**
     * Number of global state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    get globalNumByteSlice() {
        throw new NoImplementation();
    },
    /**
     * Number of local state integers in ApplicationCall
     * Min AVM version: 3
     */
    get localNumUint() {
        throw new NoImplementation();
    },
    /**
     * Number of local state byteslices in ApplicationCall
     * Min AVM version: 3
     */
    get localNumByteSlice() {
        throw new NoImplementation();
    },
    /**
     * Number of additional pages for each of the application's approval and clear state programs. An ExtraProgramPages of 1 means 2048 more total bytes, or 1024 for each program.
     * Min AVM version: 4
     */
    get extraProgramPages() {
        throw new NoImplementation();
    },
    /**
     * Marks an account nonparticipating for rewards
     * Min AVM version: 5
     */
    get nonparticipation() {
        throw new NoImplementation();
    },
    /**
     * Log messages emitted by an application call (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    logs(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Logs (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get numLogs() {
        throw new NoImplementation();
    },
    /**
     * Asset ID allocated by the creation of an ASA (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get createdAssetId() {
        throw new NoImplementation();
    },
    /**
     * ApplicationID allocated by the creation of an application (only with `itxn` in v5). Application mode only
     * Min AVM version: 5
     */
    get createdApplicationId() {
        throw new NoImplementation();
    },
    /**
     * The last message emitted. Empty bytes if none were emitted. Application mode only
     * Min AVM version: 6
     */
    get lastLog() {
        throw new NoImplementation();
    },
    /**
     * State proof public key
     * Min AVM version: 6
     */
    get stateProofPk() {
        throw new NoImplementation();
    },
    /**
     * Approval Program as an array of pages
     * Min AVM version: 7
     */
    approvalProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Number of Approval Program pages
     * Min AVM version: 7
     */
    get numApprovalProgramPages() {
        throw new NoImplementation();
    },
    /**
     * ClearState Program as an array of pages
     * Min AVM version: 7
     */
    clearStateProgramPages(a) {
        throw new NoImplementation();
    },
    /**
     * Number of ClearState Program pages
     * Min AVM version: 7
     */
    get numClearStateProgramPages() {
        throw new NoImplementation();
    },
    /**
     * Application version for which the txn must reject
     * Min AVM version: 12
     */
    get rejectVersion() {
        throw new NoImplementation();
    },
};
const VoterParams = {
    /**
     * Online stake in microalgos
     * Min AVM version: 11
     */
    voterBalance(a) {
        throw new NoImplementation();
    },
    /**
     * Had this account opted into block payouts
     * Min AVM version: 11
     */
    voterIncentiveEligible(a) {
        throw new NoImplementation();
    },
};
/**
 * Verify the proof B of message A against pubkey C. Returns vrf output and verification flag.
 * `VrfAlgorand` is the VRF used in Algorand. It is ECVRF-ED25519-SHA512-Elligator2, specified in the IETF internet draft [draft-irtf-cfrg-vrf-03](https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/03/).
 * @see Native TEAL opcode: [`vrf_verify`](https://dev.algorand.co/reference/algorand-teal/opcodes#vrf_verify)
 * Min AVM version: 7
 */
function vrfVerify(s, a, b, c) {
    throw new NoImplementation();
}
function extract(a, b, c) {
    throw new NoImplementation();
}
function select(a, b, c) {
    throw new NoImplementation();
}
function setBit(target, n, c) {
    throw new NoImplementation();
}

var op = /*#__PURE__*/Object.freeze({
    __proto__: null,
    AcctParams: AcctParams,
    AppGlobal: AppGlobal,
    AppLocal: AppLocal,
    AppParams: AppParams,
    AssetHolding: AssetHolding,
    AssetParams: AssetParams,
    get Base64 () { return Base64; },
    Block: Block,
    Box: Box,
    get Ec () { return Ec; },
    get Ecdsa () { return Ecdsa; },
    EllipticCurve: EllipticCurve,
    GITxn: GITxn,
    GTxn: GTxn,
    Global: Global,
    ITxn: ITxn,
    ITxnCreate: ITxnCreate,
    JsonRef: JsonRef,
    get MimcConfigurations () { return MimcConfigurations; },
    Scratch: Scratch,
    Txn: Txn,
    VoterParams: VoterParams,
    get VrfVerify () { return VrfVerify; },
    addw: addw,
    appOptedIn: appOptedIn,
    arg: arg,
    balance: balance,
    base64Decode: base64Decode,
    bitLength: bitLength,
    bsqrt: bsqrt,
    btoi: btoi,
    bzero: bzero,
    concat: concat,
    divmodw: divmodw,
    divw: divw,
    ecdsaPkDecompress: ecdsaPkDecompress,
    ecdsaPkRecover: ecdsaPkRecover,
    ecdsaVerify: ecdsaVerify,
    ed25519verify: ed25519verify,
    ed25519verifyBare: ed25519verifyBare,
    exp: exp,
    expw: expw,
    extract: extract,
    extractUint16: extractUint16,
    extractUint32: extractUint32,
    extractUint64: extractUint64,
    falconVerify: falconVerify,
    gaid: gaid,
    getBit: getBit,
    getByte: getByte,
    gloadBytes: gloadBytes,
    gloadUint64: gloadUint64,
    itob: itob,
    keccak256: keccak256,
    len: len,
    mimc: mimc,
    minBalance: minBalance,
    mulw: mulw,
    onlineStake: onlineStake,
    replace: replace,
    select: select,
    setBit: setBit,
    setByte: setByte,
    sha256: sha256,
    sha3_256: sha3_256,
    sha512_256: sha512_256,
    shl: shl,
    shr: shr,
    sqrt: sqrt,
    substring: substring,
    sumhash512: sumhash512,
    vrfVerify: vrfVerify
});

export { Scratch as $, AcctParams as A, Base64 as B, ed25519verifyBare as C, exp as D, Ec as E, expw as F, Global as G, extractUint16 as H, extractUint32 as I, extractUint64 as J, falconVerify as K, gaid as L, MimcConfigurations as M, getBit as N, getByte as O, GITxn as P, gloadBytes as Q, gloadUint64 as R, GTxn as S, Txn as T, itob as U, VrfVerify as V, ITxn as W, ITxnCreate as X, JsonRef as Y, keccak256 as Z, len as _, Ecdsa as a, mimc as a0, minBalance as a1, mulw as a2, onlineStake as a3, replace as a4, setByte as a5, sha256 as a6, sha3_256 as a7, sha512_256 as a8, shl as a9, shr as aa, sqrt as ab, substring as ac, sumhash512 as ad, VoterParams as ae, vrfVerify as af, extract as ag, select as ah, setBit as ai, addw as b, AppGlobal as c, AppLocal as d, appOptedIn as e, AppParams as f, arg as g, AssetHolding as h, AssetParams as i, balance as j, base64Decode as k, bitLength as l, Block as m, Box as n, op as o, bsqrt as p, btoi as q, bzero as r, concat as s, divmodw as t, divw as u, EllipticCurve as v, ecdsaPkDecompress as w, ecdsaPkRecover as x, ecdsaVerify as y, ed25519verify as z };
